// @sanity/desk-tool/lib/panes/documentPane/timeline/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatTimelineEventLabel = formatTimelineEventLabel;
exports.getTimelineEventIconComponent = getTimelineEventIconComponent;
exports.sinceTimelineProps = sinceTimelineProps;
exports.revTimelineProps = revTimelineProps;

var _closeIcon = _interopRequireDefault(require("part:@sanity/base/close-icon"));

var _editIcon = _interopRequireDefault(require("part:@sanity/base/edit-icon"));

var _plusIcon = _interopRequireDefault(require("part:@sanity/base/plus-icon"));

var _publishIcon = _interopRequireDefault(require("part:@sanity/base/publish-icon"));

var _trashIcon = _interopRequireDefault(require("part:@sanity/base/trash-icon"));

var _unpublishIcon = _interopRequireDefault(require("part:@sanity/base/unpublish-icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LABELS = {
  create: 'created',
  delete: 'deleted',
  discardDraft: 'discarded draft',
  initial: 'created',
  editDraft: 'edited',
  publish: 'published',
  unpublish: 'unpublished'
};
var ICON_COMPONENTS = {
  create: _plusIcon.default,
  delete: _trashIcon.default,
  discardDraft: _closeIcon.default,
  initial: _plusIcon.default,
  editDraft: _editIcon.default,
  publish: _publishIcon.default,
  unpublish: _unpublishIcon.default
};

function formatTimelineEventLabel(type) {
  return LABELS[type];
}

function getTimelineEventIconComponent(type) {
  return ICON_COMPONENTS[type];
}

function sinceTimelineProps(since, rev) {
  return {
    topSelection: rev,
    bottomSelection: since,
    disabledBeforeSelection: true
  };
}

function revTimelineProps(rev) {
  return {
    topSelection: rev,
    bottomSelection: rev
  };
}
