// @sanity/components/lib/menuButton/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _menuButton = require("part:@sanity/components/menu-button");

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  var open = (0, _knobs.boolean)('Open', false, 'Props');
  var placement = (0, _knobs.select)('Placement', {
    'top-start': 'Top start',
    top: 'Top',
    'top-end': 'Top end',
    'right-start': 'Right start',
    right: 'Right',
    'right-end': 'Right end',
    'bottom-start': 'Bottom start',
    bottom: 'Bottom',
    'bottom-end': 'Bottom end',
    'left-start': 'Left start',
    left: 'Left',
    'left-end': 'Left end'
  }, 'bottom', 'Props');
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/dialogs/default",
    propTables: [_menuButton.MenuButton]
  }, /*#__PURE__*/_react.default.createElement(_menuButton.MenuButton, {
    menu: /*#__PURE__*/_react.default.createElement("div", {
      style: {
        padding: 20
      }
    }, "This is the menu"),
    open: open,
    placement: placement,
    setOpen: (0, _actions.action)('setOpen')
  }, "Open menu")));
}
