// @sanity/components/lib/edititem/stories/fold.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FoldStory = FoldStory;

var _chance = _interopRequireDefault(require("chance"));

var _fold = _interopRequireDefault(require("part:@sanity/components/edititem/fold"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireDefault(require("react"));

var _components = require("part:@sanity/storybook/components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();
var wrapperStyles = {
  width: '50%',
  margin: '0 auto',
  backgroundColor: '#ccc',
  minHeight: '50vh',
  maxHeight: '70vh',
  paddingTop: '5rem',
  position: 'relative',
  overflow: 'scroll'
};

function FoldStory() {
  return /*#__PURE__*/_react.default.createElement(_components.Container, null, /*#__PURE__*/_react.default.createElement("div", {
    style: wrapperStyles
  }, /*#__PURE__*/_react.default.createElement("p", null, "Over"), /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/edititem/fold",
    propTables: [_fold.default]
  }, /*#__PURE__*/_react.default.createElement(_fold.default, {
    title: "Edit this item",
    onClose: (0, _actions.action)('onClose')
  }, (0, _knobs.text)('children (prop)', 'Put your content here'), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      height: "".concat((0, _knobs.number)('content padding', 10), "px")
    }
  }))), /*#__PURE__*/_react.default.createElement("p", null, "Under"), chance.paragraph(), chance.paragraph(), chance.paragraph(), chance.paragraph(), chance.paragraph(), chance.paragraph(), chance.paragraph(), chance.paragraph()));
}
