// @sanity/portable-text-editor/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
var PortableTextEditor_1 = require("./editor/PortableTextEditor");
Object.defineProperty(exports, "PortableTextEditor", { enumerable: true, get: function () { return PortableTextEditor_1.PortableTextEditor; } });
Object.defineProperty(exports, "keyGenerator", { enumerable: true, get: function () { return PortableTextEditor_1.defaultKeyGenerator; } });
__exportStar(require("./types/editor"), exports);
__exportStar(require("./types/portableText"), exports);
__exportStar(require("./types/schema"), exports);
__exportStar(require("./types/options"), exports);
var patches_1 = require("./utils/patches");
Object.defineProperty(exports, "compactPatches", { enumerable: true, get: function () { return patches_1.compactPatches; } });
var getPortableTextFeatures_1 = require("./utils/getPortableTextFeatures");
Object.defineProperty(exports, "getPortableTextFeatures", { enumerable: true, get: function () { return getPortableTextFeatures_1.getPortableTextFeatures; } });
var Editable_1 = require("./editor/Editable");
Object.defineProperty(exports, "PortableTextEditable", { enumerable: true, get: function () { return Editable_1.PortableTextEditable; } });
var usePortableTextEditor_1 = require("./editor/hooks/usePortableTextEditor");
Object.defineProperty(exports, "usePortableTextEditor", { enumerable: true, get: function () { return usePortableTextEditor_1.usePortableTextEditor; } });
var usePortableTextEditorSelection_1 = require("./editor/hooks/usePortableTextEditorSelection");
Object.defineProperty(exports, "usePortableTextEditorSelection", { enumerable: true, get: function () { return usePortableTextEditorSelection_1.usePortableTextEditorSelection; } });
