// @sanity/desk-tool/lib/components/ConfirmUnpublish.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _alert = _interopRequireDefault(require("part:@sanity/components/alerts/alert"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _spinner = _interopRequireDefault(require("part:@sanity/components/loading/spinner"));

var _enhanceWithReferringDocuments = _interopRequireDefault(require("./enhanceWithReferringDocuments"));

var _DocTitle = _interopRequireDefault(require("./DocTitle"));

var _ReferringDocumentsList = _interopRequireDefault(require("./ReferringDocumentsList"));

var _ConfirmUnpublish = _interopRequireDefault(require("./ConfirmUnpublish.css"));

var _class, _temp;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var _default = (0, _enhanceWithReferringDocuments.default)((_temp = _class = class ConfirmUnpublish extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleAction", action => {
      var _this$props = this.props,
          onCancel = _this$props.onCancel,
          onConfirm = _this$props.onConfirm;

      if (action.name === 'confirm') {
        onConfirm();
      }

      if (action.name === 'cancel') {
        onCancel();
      }
    });
  }

  // eslint-disable-next-line complexity
  render() {
    var _this$props2 = this.props,
        isCheckingReferringDocuments = _this$props2.isCheckingReferringDocuments,
        referringDocuments = _this$props2.referringDocuments,
        draft = _this$props2.draft,
        published = _this$props2.published,
        onCancel = _this$props2.onCancel;
    var hasReferringDocuments = referringDocuments.length > 0;
    var canContinue = !isCheckingReferringDocuments;
    var actions = [canContinue && {
      color: 'danger',
      name: 'confirm',
      title: hasReferringDocuments ? 'Try to unpublish anyway' : 'Unpublish now'
    }, {
      inverted: true,
      name: 'cancel',
      title: 'Cancel'
    }].filter(Boolean);
    var title = isCheckingReferringDocuments ? 'Checking…' : 'Confirm unpublish';

    var docTitle = /*#__PURE__*/_react.default.createElement(_DocTitle.default, {
      document: draft || published
    });

    return /*#__PURE__*/_react.default.createElement(_fullscreen.default, {
      cardClassName: _ConfirmUnpublish.default.card,
      isOpen: true,
      showHeader: true,
      centered: true,
      title: title,
      onClose: onCancel,
      onAction: this.handleAction,
      actions: actions
    }, isCheckingReferringDocuments && /*#__PURE__*/_react.default.createElement(_spinner.default, {
      message: "Looking for referring documents\u2026"
    }), hasReferringDocuments && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_alert.default, {
      color: "warning",
      icon: _warningIcon.default
    }, "Warning: Found", ' ', referringDocuments.length === 1 ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "a document") : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, referringDocuments.length, " documents"), ' ', "that refers to \u201C", docTitle, "\u201D"), /*#__PURE__*/_react.default.createElement("p", null, "You may not be able to unpublish \u201C", docTitle, "\u201D because the following document", referringDocuments.length !== 1 && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "s"), " refers to it:"), /*#__PURE__*/_react.default.createElement(_ReferringDocumentsList.default, {
      documents: referringDocuments
    })), !isCheckingReferringDocuments && !hasReferringDocuments && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_alert.default, {
      color: "warning",
      icon: _warningIcon.default,
      title: "Careful!"
    }, "If you unpublish this document, it will no longer be available for the public. However, it will not be deleted and can be published again later."), /*#__PURE__*/_react.default.createElement("p", null, "Are you sure you want to unpublish the document ", /*#__PURE__*/_react.default.createElement("strong", null, "\u201C", docTitle, "\u201D"), "?")));
  }

}, _defineProperty(_class, "propTypes", {
  onCancel: _propTypes.default.func.isRequired,
  onConfirm: _propTypes.default.func.isRequired,
  published: _propTypes.default.object,
  draft: _propTypes.default.object,
  referringDocuments: _propTypes.default.array,
  isCheckingReferringDocuments: _propTypes.default.bool
}), _temp));

exports.default = _default;
