// @sanity/desk-tool/lib/components/Diff.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _diff = require("diff");

var _Diff = _interopRequireDefault(require("./Diff.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function getDiffStatKey(part) {
  if (part.added) {
    return 'added';
  }

  if (part.removed) {
    return 'removed';
  }

  return 'neutral';
}

class Diff extends _react.default.PureComponent {
  render() {
    var diff = (0, _diff.diffJson)(this.props.inputA, this.props.inputB);
    return /*#__PURE__*/_react.default.createElement("pre", null, diff.map((part, index) => /*#__PURE__*/_react.default.createElement("span", {
      key: index,
      className: _Diff.default[getDiffStatKey(part)]
    }, part.value)));
  }

}

exports.default = Diff;

_defineProperty(Diff, "defaultProps", {
  inputA: '',
  inputB: ''
});

_defineProperty(Diff, "propTypes", {
  inputA: _propTypes.default.object,
  inputB: _propTypes.default.object
});
