// @sanity/desk-tool/lib/components/DocTitle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DocTitle;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _schema = _interopRequireDefault(require("part:@sanity/base/schema"));

var _preview = require("part:@sanity/base/preview");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/no-multi-comp */
function ShowTitle(_ref) {
  var title = _ref.title;
  return /*#__PURE__*/_react.default.createElement("span", null, title);
}

ShowTitle.propTypes = {
  title: _propTypes.default.string
};

function DocTitle(props) {
  var document = props.document;

  var type = _schema.default.get(document._type);

  return /*#__PURE__*/_react.default.createElement(_preview.PreviewFields, {
    document: document,
    type: type,
    fields: ['title']
  }, ShowTitle);
}

DocTitle.propTypes = {
  document: _propTypes.default.shape({
    _type: _propTypes.default.string
  })
};
