// @sanity/desk-tool/lib/components/MissingDocumentTypesMessage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = MissingDocumentTypesMessage;

var _react = _interopRequireDefault(require("react"));

var _MissingDocumentTypesMessage = _interopRequireDefault(require("./MissingDocumentTypesMessage.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function MissingDocumentTypesMessage(props) {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _MissingDocumentTypesMessage.default.root
  }, /*#__PURE__*/_react.default.createElement("h2", null, "Empty schema"), /*#__PURE__*/_react.default.createElement("p", null, "Your schema does not contain any document types. If it did, those types would be listed here.", ' ', /*#__PURE__*/_react.default.createElement("a", {
    title: "Schema documentation",
    target: "_blank",
    rel: "noopener noreferrer",
    href: "https://www.sanity.io/docs/content-studio/the-schema"
  }, "Read more about how to add schema types"), "."));
}
