// @sanity/desk-tool/lib/components/MissingSchemaType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _preview = require("part:@sanity/base/preview");

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getUnknownTypeFallback = (id, typeName) => ({
  title: /*#__PURE__*/_react.default.createElement("span", {
    style: {
      fontStyle: 'italic'
    }
  }, "No schema found for type \"", typeName, "\""),
  subtitle: /*#__PURE__*/_react.default.createElement("span", {
    style: {
      fontStyle: 'italic'
    }
  }, "Document: ", id),
  media: _warningIcon.default
});

class MissingSchemaType extends _react.default.Component {
  render() {
    var _this$props = this.props,
        layout = _this$props.layout,
        value = _this$props.value;
    return /*#__PURE__*/_react.default.createElement(_preview.SanityDefaultPreview, {
      value: getUnknownTypeFallback(value._id, value._type),
      layout: layout
    });
  }

}

exports.default = MissingSchemaType;
MissingSchemaType.propTypes = {
  layout: _propTypes.default.string,
  value: _propTypes.default.object
};
