// @sanity/desk-tool/lib/components/ReferringDocumentsList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ReferringDocumentsList;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _default = require("part:@sanity/components/lists/default");

var _preview = _interopRequireDefault(require("part:@sanity/base/preview"));

var _router = require("part:@sanity/base/router");

var _schema = _interopRequireDefault(require("part:@sanity/base/schema"));

var _ReferringDocumentsList = _interopRequireDefault(require("./ReferringDocumentsList.css"));

var _DraftStatus = _interopRequireDefault(require("./DraftStatus"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ReferringDocumentsList(props) {
  var documents = props.documents;
  return /*#__PURE__*/_react.default.createElement(_default.List, {
    className: _ReferringDocumentsList.default.root
  }, documents.map(document => {
    var schemaType = _schema.default.get(document._type);

    return /*#__PURE__*/_react.default.createElement(_default.Item, {
      className: _ReferringDocumentsList.default.item,
      key: document._id
    }, schemaType ? /*#__PURE__*/_react.default.createElement(_router.IntentLink, {
      className: _ReferringDocumentsList.default.link,
      intent: "edit",
      params: {
        id: document._id,
        type: document._type
      }
    }, /*#__PURE__*/_react.default.createElement(_preview.default, {
      value: document,
      type: schemaType
    }), document._hasDraft && /*#__PURE__*/_react.default.createElement(_DraftStatus.default, null)) : /*#__PURE__*/_react.default.createElement("div", null, "A document of the unknown type ", /*#__PURE__*/_react.default.createElement("em", null, document._type)));
  }));
}

ReferringDocumentsList.propTypes = {
  documents: _propTypes.default.arrayOf(_propTypes.default.shape({
    _id: _propTypes.default.string,
    _type: _propTypes.default.string
  }))
};
