// @sanity/form-builder/lib/inputs/ArrayInput/ArrayFunctions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _types = require("@sanity/types");

var _classnames = _interopRequireDefault(require("classnames"));

var _dropdown = _interopRequireDefault(require("part:@sanity/components/buttons/dropdown"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _buttonGrid = _interopRequireDefault(require("part:@sanity/components/buttons/button-grid"));

var _plusIcon = _interopRequireDefault(require("part:@sanity/base/plus-icon"));

var _react = _interopRequireDefault(require("react"));

var _ArrayFunctions = _interopRequireDefault(require("./ArrayFunctions.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class ArrayFunctions extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleDropDownAction", menuItem => {
      this.handleInsertItem(menuItem.type);
    });

    _defineProperty(this, "handleAddBtnClick", () => {
      this.handleInsertItem(this.props.type.of[0]);
    });

    _defineProperty(this, "handleInsertItem", type => {
      var _this$props = this.props,
          onCreateValue = _this$props.onCreateValue,
          onAppendItem = _this$props.onAppendItem;
      var item = onCreateValue(type);
      onAppendItem(item);
    });
  }

  renderSelectType() {
    var items = this.props.type.of.map(memberDef => {
      // Use reference icon if reference is to one type only
      var referenceIcon = (0, _types.isReferenceSchemaType)(memberDef) && (memberDef.to || []).length === 1 && memberDef.to[0].icon;
      var icon = memberDef.icon || memberDef.type.icon || referenceIcon || _plusIcon.default;
      return {
        title: memberDef.title || memberDef.type.name,
        type: memberDef,
        icon
      };
    });
    return /*#__PURE__*/_react.default.createElement(_dropdown.default, {
      inverted: true,
      items: items,
      onAction: this.handleDropDownAction
    }, "Add");
  }

  render() {
    var _this$props2 = this.props,
        className = _this$props2.className,
        type = _this$props2.type,
        readOnly = _this$props2.readOnly,
        children = _this$props2.children;

    if (readOnly) {
      return null;
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)(_ArrayFunctions.default.root, className)
    }, /*#__PURE__*/_react.default.createElement(_buttonGrid.default, {
      align: "start"
    }, type.of.length === 1 ? /*#__PURE__*/_react.default.createElement(_default.default, {
      inverted: true,
      onClick: this.handleAddBtnClick
    }, "Add") : this.renderSelectType(), children || null));
  }

}

exports.default = ArrayFunctions;
