// @sanity/components/lib/clickOutside/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _clickOutside = require("part:@sanity/components/click-outside");

var _actions = require("part:@sanity/storybook/addons/actions");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/buttons/default",
    propTables: [_clickOutside.ClickOutside]
  }, /*#__PURE__*/_react.default.createElement(_clickOutside.ClickOutside, {
    onClickOutside: (0, _actions.action)('click outside')
  }, ref => /*#__PURE__*/_react.default.createElement("div", {
    onClick: (0, _actions.action)('click inside'),
    ref: ref,
    style: {
      background: '#fff',
      border: '1px solid #999',
      padding: '2em'
    }
  }, "click inside/outside"))));
}
