// @sanity/components/lib/radiobutton/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _default = _interopRequireDefault(require("part:@sanity/components/radiobutton/default"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var centerStyle = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  height: '100%',
  width: '100%',
  position: 'absolute',
  top: 0,
  left: 0
};

function DefaultStory() {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/radiobutton/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    name: "radioButton",
    label: (0, _knobs.text)('label', 'Label', 'props'),
    item: (0, _knobs.object)('Item', {
      title: 'test'
    }, 'props'),
    checked: (0, _knobs.boolean)('checked', false, 'props'),
    onChange: (0, _actions.action)('onChange')
  })));
}
