// @sanity/components/lib/dialogs/PanePopover.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _circleCheckIcon = _interopRequireDefault(require("part:@sanity/base/circle-check-icon"));

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _infoIcon = _interopRequireDefault(require("part:@sanity/base/info-icon"));

var _classnames = _interopRequireDefault(require("classnames"));

var _PanePopover = _interopRequireDefault(require("./PanePopover.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DEFAULT_ICONS = {
  info: /*#__PURE__*/_react.default.createElement(_infoIcon.default, null),
  success: /*#__PURE__*/_react.default.createElement(_circleCheckIcon.default, null),
  warning: /*#__PURE__*/_react.default.createElement(_warningIcon.default, null),
  error: /*#__PURE__*/_react.default.createElement(_warningIcon.default, null)
}; // @todo: refactor to functional component

class PanePopover extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "iconKind", () => {
      var _this$props = this.props,
          icon = _this$props.icon,
          _this$props$kind = _this$props.kind,
          kind = _this$props$kind === void 0 ? 'info' : _this$props$kind;
      if (kind && typeof icon === 'boolean' && icon) return DEFAULT_ICONS[kind];
      if (typeof icon === 'object') return icon;
      return undefined;
    });
  }

  render() {
    var _this$props2 = this.props,
        children = _this$props2.children,
        icon = _this$props2.icon,
        id = _this$props2.id,
        _this$props2$kind = _this$props2.kind,
        kind = _this$props2$kind === void 0 ? 'info' : _this$props2$kind,
        title = _this$props2.title,
        subtitle = _this$props2.subtitle;
    var iconNode = this.iconKind();
    return /*#__PURE__*/_react.default.createElement("div", {
      "aria-label": kind,
      "aria-describedby": "popoverTitle-".concat(kind, "-").concat(id),
      className: (0, _classnames.default)(_PanePopover.default.root, _PanePopover.default.dialog),
      "data-kind": kind
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _PanePopover.default.inner
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _PanePopover.default.content
    }, /*#__PURE__*/_react.default.createElement("div", {
      id: "popoverTitle-".concat(kind, "-").concat(id),
      className: _PanePopover.default.title
    }, icon && /*#__PURE__*/_react.default.createElement("div", {
      role: "img",
      "aria-hidden": true,
      className: _PanePopover.default.icon
    }, iconNode), title), subtitle && /*#__PURE__*/_react.default.createElement("div", {
      className: _PanePopover.default.subtitle
    }, subtitle), children && /*#__PURE__*/_react.default.createElement("div", {
      className: _PanePopover.default.children
    }, children))));
  }

}

exports.default = PanePopover;
