// @sanity/form-builder/lib/inputs/ImageInput/AssetMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = AssetMenu;

var _linkIcon = _interopRequireDefault(require("part:@sanity/base/link-icon"));

var _moreVertIcon = _interopRequireDefault(require("part:@sanity/base/more-vert-icon"));

var _trashIcon = _interopRequireDefault(require("part:@sanity/base/trash-icon"));

var _dropdown = _interopRequireDefault(require("part:@sanity/components/buttons/dropdown"));

var _react = _interopRequireDefault(require("react"));

var _AssetMenu = _interopRequireDefault(require("./AssetMenu.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var menuItems = [{
  name: 'showRefs',
  title: 'Show documents using this',
  icon: _linkIcon.default
}, {
  name: 'delete',
  title: 'Delete',
  color: 'danger',
  icon: _trashIcon.default
}];

function MenuItem(item) {
  var color = item.color,
      title = item.title,
      icon = item.icon;
  var Icon = icon;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: color === 'danger' ? _AssetMenu.default.menuItemDanger : _AssetMenu.default.menuItem
  }, /*#__PURE__*/_react.default.createElement("span", null, icon && /*#__PURE__*/_react.default.createElement(Icon, null)), /*#__PURE__*/_react.default.createElement("span", null, title));
}

function AssetMenu(_ref) {
  var isSelected = _ref.isSelected,
      onAction = _ref.onAction;
  return /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    icon: _moreVertIcon.default,
    padding: "small",
    placement: "bottom-end",
    showArrow: false,
    items: isSelected ? menuItems.filter(item => item.name !== 'delete') : menuItems,
    renderItem: MenuItem,
    onAction: onAction
  });
}
