// @sanity/form-builder/lib/inputs/ImageInput/DefaultSource.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _imageIcon = _interopRequireDefault(require("part:@sanity/base/image-icon"));

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _default3 = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _Asset = _interopRequireDefault(require("./Asset"));

var _DefaultSource = _interopRequireDefault(require("./DefaultSource.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var PER_PAGE = 200;

var buildQuery = function buildQuery() {
  var start = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
  var end = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : PER_PAGE;
  return "\n  *[_type == \"sanity.imageAsset\"] | order(_updatedAt desc) [".concat(start, "...").concat(end, "] {\n    _id,\n    url,\n    metadata {dimensions}\n  }\n");
};

class DefaultSource extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "state", {
      assets: [],
      isLastPage: false,
      isLoading: false
    });

    _defineProperty(this, "pageNo", 0);

    _defineProperty(this, "handleDeleteFinished", id => {
      this.setState(prevState => ({
        assets: prevState.assets.filter(asset => asset._id !== id)
      }));
    });

    _defineProperty(this, "handleItemClick", event => {
      event.preventDefault();
      this.select(event.currentTarget.getAttribute('data-id'));
    });

    _defineProperty(this, "handleItemKeyPress", event => {
      if (event.key === 'Enter') {
        event.preventDefault();
        this.select(event.currentTarget.getAttribute('data-id'));
      }
    });

    _defineProperty(this, "handleClose", () => {
      if (this.props.onClose) {
        this.props.onClose();
      }
    });

    _defineProperty(this, "handleFetchNextPage", () => {
      this.fetchPage(++this.pageNo);
    });
  }

  fetchPage(pageNo) {
    var start = pageNo * PER_PAGE;
    var end = start + PER_PAGE;
    this.setState({
      isLoading: true
    });
    return _client.default.fetch(buildQuery(start, end)).then(result => {
      this.setState(prevState => ({
        isLastPage: result.length < PER_PAGE,
        assets: prevState.assets.concat(result),
        isLoading: false
      }));
    });
  }

  componentDidMount() {
    this.fetchPage(this.pageNo);
  }

  select(id) {
    var selected = this.state.assets.find(doc => doc._id === id);

    if (selected) {
      this.props.onSelect([{
        kind: 'assetDocumentId',
        value: id
      }]);
    }
  }

  render() {
    var selectedAssets = this.props.selectedAssets;
    var _this$state = this.state,
        assets = _this$state.assets,
        isLastPage = _this$state.isLastPage,
        isLoading = _this$state.isLoading;
    return /*#__PURE__*/_react.default.createElement(_default3.default, {
      isOpen: true,
      onClose: this.handleClose,
      size: "large",
      title: "Select image"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _DefaultSource.default.imageGrid
    }, assets.map(asset => /*#__PURE__*/_react.default.createElement(_Asset.default, {
      key: asset._id,
      asset: asset,
      isSelected: selectedAssets.some(selected => selected._id === asset._id),
      onClick: this.handleItemClick,
      onKeyPress: this.handleItemKeyPress,
      onDeleteFinished: this.handleDeleteFinished
    }))), !isLoading && assets.length === 0 && /*#__PURE__*/_react.default.createElement("div", {
      className: _DefaultSource.default.noAssets
    }, "No images found"), /*#__PURE__*/_react.default.createElement("div", {
      className: _DefaultSource.default.loadMore
    }, !isLastPage && /*#__PURE__*/_react.default.createElement(_default2.default, {
      inverted: true,
      onClick: this.handleFetchNextPage,
      loading: isLoading
    }, "Load more")));
  }

}

var _default = {
  name: 'sanity-default',
  title: 'Uploaded images',
  component: DefaultSource,
  icon: _imageIcon.default
};
exports.default = _default;
