// @sanity/default-login/lib/ErrorDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ErrorDialog;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _fullscreenMessage = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen-message"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _CorsCheck = _interopRequireDefault(require("./CorsCheck"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ErrorDialog(props) {
  var isNetworkError = props.error.isNetworkError;
  return /*#__PURE__*/_react.default.createElement(_fullscreenMessage.default, {
    buttons: /*#__PURE__*/_react.default.createElement(_default.default, {
      onClick: props.onRetry
    }, "Retry"),
    color: "danger",
    title: "Error",
    isOpen: true,
    centered: true
  }, !isNetworkError && /*#__PURE__*/_react.default.createElement("p", null, props.error.message), isNetworkError && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, "An error occured while attempting to reach the Sanity API."), /*#__PURE__*/_react.default.createElement(_CorsCheck.default, null)));
}

ErrorDialog.propTypes = {
  error: _propTypes.default.shape({
    isNetworkError: _propTypes.default.bool,
    message: _propTypes.default.string.isRequired
  }).isRequired,
  onRetry: _propTypes.default.func.isRequired
};
