// @sanity/default-login/lib/UnauthorizedUser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = UnauthorizedUser;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _fullscreenMessage = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen-message"));

var _user = _interopRequireDefault(require("part:@sanity/base/user"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function handleLogout() {
  _user.default.actions.logout();
}

function UnauthorizedUser(props) {
  return /*#__PURE__*/_react.default.createElement(_fullscreenMessage.default, {
    buttons: /*#__PURE__*/_react.default.createElement(_default.default, {
      onClick: handleLogout
    }, "Logout"),
    title: "Unauthorized"
  }, /*#__PURE__*/_react.default.createElement("p", null, "You are not authorized to access this studio. Maybe you could ask someone to invite you to collaborate on this project?"), /*#__PURE__*/_react.default.createElement("p", null, "If you think this is an error, verify that you are logged in with the correct account. You are currently logged in as", ' ', /*#__PURE__*/_react.default.createElement("span", null, props.user.name, " (", props.user.email, ")")));
}

UnauthorizedUser.propTypes = {
  user: _propTypes.default.shape({
    name: _propTypes.default.string.isRequired,
    email: _propTypes.default.string.isRequired
  })
};
