// @sanity/components/lib/textinputs/stories/customStyle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CustomStyleStory = CustomStyleStory;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/textinputs/default"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _customStyle = _interopRequireDefault(require("./customStyle.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function CustomStyleStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 640
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/textinputs/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    placeholder: (0, _knobs.text)('placeholder', 'This is the placeholder', 'props'),
    value: (0, _knobs.text)('value', false, 'props'),
    type: (0, _knobs.select)('type', ['text', 'number', 'email', 'tel'], 'text', 'props'),
    selected: (0, _knobs.boolean)('selected', false, 'props'),
    onChange: (0, _actions.action)('onChange'),
    onFocus: (0, _actions.action)('onFocus'),
    onKeyPress: (0, _actions.action)('onKeyPress'),
    onBlur: (0, _actions.action)('onBlur'),
    styles: (0, _knobs.object)('styles', _customStyle.default, 'props'),
    id: "ThisIsAnUniqueId"
  }))));
}
