// @sanity/components/lib/textinputs/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/textinputs/default"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 640
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/textinputs/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    disabled: (0, _knobs.boolean)('Disabled', false, 'props'),
    hasError: (0, _knobs.boolean)('Has errors', false, 'props'),
    id: "ThisIsAnUniqueId",
    isSelected: (0, _knobs.boolean)('Is selected', false, 'props'),
    onChange: (0, _actions.action)('onChange'),
    onFocus: (0, _actions.action)('onFocus'),
    onKeyPress: (0, _actions.action)('onKeyPress'),
    onBlur: (0, _actions.action)('onBlur'),
    placeholder: (0, _knobs.text)('Placeholder', 'This is the placeholder', 'props'),
    readOnly: (0, _knobs.boolean)('Read-only', false, 'props'),
    type: (0, _knobs.select)('Type', ['text', 'number', 'email', 'tel'], 'text', 'props'),
    value: (0, _knobs.text)('Value', '', 'props')
  }))));
}
