// @sanity/field/lib/diff/components/DiffInspectWrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DiffInspectWrapper = DiffInspectWrapper;

var _react = _interopRequireWildcard(require("react"));

var _paths = require("../../paths");

var _DiffInspectWrapper = _interopRequireDefault(require("./DiffInspectWrapper.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function DiffInspectWrapper(_ref) {
  var children = _ref.children,
      className = _ref.className,
      change = _ref.change;
  var isHovering = (0, _react.useRef)(false);

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      isInspecting = _useState2[0],
      setIsInspecting = _useState2[1];

  var toggleInspect = (0, _react.useCallback)(() => setIsInspecting(state => !state), [setIsInspecting]);
  var handleMouseEnter = (0, _react.useCallback)(() => isHovering.current = true, []);
  var handleMouseLeave = (0, _react.useCallback)(() => isHovering.current = false, [isHovering]);
  (0, _react.useEffect)(() => {
    function onKeyDown(evt) {
      var metaKey = evt.metaKey,
          key = evt.key;

      if (metaKey && key === 'i' && isHovering.current) {
        toggleInspect();
      }
    }

    window.addEventListener('keydown', onKeyDown, false);
    return () => window.removeEventListener('keydown', onKeyDown, false);
  }, [toggleInspect]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: className,
    onMouseEnter: handleMouseEnter,
    onMouseLeave: handleMouseLeave
  }, isInspecting ? /*#__PURE__*/_react.default.createElement(DiffInspector, {
    change: change
  }) : children);
}

function DiffInspector(_ref2) {
  var _change$itemDiff, _change$itemDiff2, _change$itemDiff3;

  var change = _ref2.change;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: _DiffInspectWrapper.default.meta
  }, printMeta({
    path: (0, _paths.pathToString)(change.path),
    fromIndex: (_change$itemDiff = change.itemDiff) === null || _change$itemDiff === void 0 ? void 0 : _change$itemDiff.fromIndex,
    toIndex: (_change$itemDiff2 = change.itemDiff) === null || _change$itemDiff2 === void 0 ? void 0 : _change$itemDiff2.toIndex,
    hasMoved: (_change$itemDiff3 = change.itemDiff) === null || _change$itemDiff3 === void 0 ? void 0 : _change$itemDiff3.hasMoved,
    action: change.diff.action,
    isChanged: change.diff.isChanged
  })), /*#__PURE__*/_react.default.createElement("pre", {
    className: _DiffInspectWrapper.default.fromJson
  }, jsonify(change.diff.fromValue)), /*#__PURE__*/_react.default.createElement("div", {
    className: _DiffInspectWrapper.default.arrow
  }, "\u2193"), /*#__PURE__*/_react.default.createElement("pre", {
    className: _DiffInspectWrapper.default.toJson
  }, jsonify(change.diff.toValue)));
}

function jsonify(value) {
  if (typeof value === 'undefined') {
    return 'undefined';
  }

  return JSON.stringify(value, null, 2);
}

function printMeta(keys) {
  var lines = [];
  Object.entries(keys).forEach((_ref3) => {
    var _ref4 = _slicedToArray(_ref3, 2),
        key = _ref4[0],
        value = _ref4[1];

    if (typeof value !== 'undefined' && value !== null) {
      lines.push("".concat(key, ": ").concat(value));
    }
  });
  return lines.join('\n');
}
