// @sanity/field/lib/diff/components/DiffErrorBoundary.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DiffErrorBoundary = void 0;

var React = _interopRequireWildcard(require("react"));

var _errorOutlineIcon = _interopRequireDefault(require("part:@sanity/base/error-outline-icon"));

var _ValueError = _interopRequireDefault(require("./ValueError.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

class DiffErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      error: undefined
    };
  }

  static getDerivedStateFromError(error) {
    return {
      error
    };
  } // eslint-disable-next-line class-methods-use-this


  componentDidCatch(error) {
    /* eslint-disable no-console */
    console.error('Error rendering diff component: ');
    console.error(error);
    /* eslint-enable no-console */
  }

  render() {
    var isDev = __DEV__ === true;
    var error = this.state.error;

    if (!error) {
      return this.props.children;
    }

    var help = isDev ? /*#__PURE__*/React.createElement("p", null, "Check the developer console for more information.") : null;
    return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("div", {
      className: _ValueError.default.root
    }, /*#__PURE__*/React.createElement("span", {
      className: _ValueError.default.icon
    }, /*#__PURE__*/React.createElement(_errorOutlineIcon.default, null)), /*#__PURE__*/React.createElement("div", {
      className: _ValueError.default.message
    }, "The component responsible for showing the changes to this field has crashed.", help)));
  }

}

exports.DiffErrorBoundary = DiffErrorBoundary;
