// @sanity/field/lib/diff/components/DiffTooltip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DiffTooltip = DiffTooltip;

var _components = require("@sanity/base/components");

var _hooks = require("@sanity/base/hooks");

var _react = _interopRequireDefault(require("react"));

var _tooltip = require("part:@sanity/components/tooltip");

var _diff = require("../../diff");

var _annotations = require("../annotations");

var _DiffTooltip = _interopRequireDefault(require("./DiffTooltip.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function DiffTooltip(props) {
  if ('diff' in props) {
    var diff = props.diff,
        _props$path = props.path,
        path = _props$path === void 0 ? [] : _props$path,
        restProps = _objectWithoutProperties(props, ["diff", "path"]);

    var annotation = (0, _diff.getAnnotationAtPath)(diff, path);
    return /*#__PURE__*/_react.default.createElement(DiffTooltipWithAnnotation, _extends({}, restProps, {
      annotations: annotation ? [annotation] : []
    }));
  }

  return /*#__PURE__*/_react.default.createElement(DiffTooltipWithAnnotation, props);
}

function DiffTooltipWithAnnotation(props) {
  var annotations = props.annotations,
      children = props.children,
      _props$description = props.description,
      description = _props$description === void 0 ? 'Changed' : _props$description,
      _props$placement = props.placement,
      placement = _props$placement === void 0 ? 'auto' : _props$placement;

  if (!annotations) {
    return children;
  }

  var content = /*#__PURE__*/_react.default.createElement("div", {
    className: _DiffTooltip.default.root
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _DiffTooltip.default.description
  }, description), annotations.map((annotation, idx) => /*#__PURE__*/_react.default.createElement(AnnotationItem, {
    annotation: annotation,
    key: idx
  })));

  return /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, {
    content: content,
    placement: placement,
    portal: true,
    allowedAutoPlacements: ['top', 'bottom']
  }, children);
}

function AnnotationItem(_ref) {
  var annotation = _ref.annotation;
  var author = annotation.author,
      timestamp = annotation.timestamp;

  var _useUser = (0, _hooks.useUser)(author),
      error = _useUser.error,
      user = _useUser.value;

  var color = (0, _annotations.useAnnotationColor)(annotation);
  var timeAgo = (0, _hooks.useTimeAgo)(timestamp, {
    minimal: true
  }); // @todo handle?

  if (error) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: _DiffTooltip.default.annotation
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _DiffTooltip.default.user,
    style: {
      backgroundColor: color.background,
      color: color.text
    }
  }, /*#__PURE__*/_react.default.createElement("span", null, /*#__PURE__*/_react.default.createElement(_components.UserAvatar, {
    userId: author
  })), /*#__PURE__*/_react.default.createElement("span", {
    className: _DiffTooltip.default.displayName
  }, user ? user.displayName : 'Loading…')), /*#__PURE__*/_react.default.createElement("time", {
    className: _DiffTooltip.default.timeAgo,
    dateTime: timestamp
  }, timeAgo));
}
