// @sanity/field/lib/diff/components/FallbackDiff.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FallbackDiff = void 0;

var _react = _interopRequireDefault(require("react"));

var _preview = _interopRequireDefault(require("part:@sanity/base/preview"));

var _DiffFromTo = require("./DiffFromTo");

var _FallbackDiff = _interopRequireDefault(require("./FallbackDiff.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FallbackPreview = (_ref) => {
  var value = _ref.value,
      schemaType = _ref.schemaType;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _FallbackDiff.default.root
  }, /*#__PURE__*/_react.default.createElement(_preview.default, {
    type: schemaType,
    value: value,
    layout: "default"
  }));
};

var FallbackDiff = (_ref2) => {
  var diff = _ref2.diff,
      schemaType = _ref2.schemaType;
  return /*#__PURE__*/_react.default.createElement(_DiffFromTo.DiffFromTo, {
    diff: diff,
    schemaType: schemaType,
    previewComponent: FallbackPreview,
    cardClassName: _FallbackDiff.default.card,
    layout: "grid"
  });
};

exports.FallbackDiff = FallbackDiff;
