// @sanity/field/lib/diff/components/FieldChange.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FieldChange = FieldChange;

var _reactHooks = require("@sanity/react-hooks");

var _classnames = _interopRequireDefault(require("classnames"));

var _popover = _interopRequireDefault(require("part:@sanity/components/dialogs/popover"));

var _react = _interopRequireWildcard(require("react"));

var _undoChange = require("../changes/undoChange");

var _DiffContext = require("../contexts/DiffContext");

var _ChangeBreadcrumb = require("./ChangeBreadcrumb");

var _DiffErrorBoundary = require("./DiffErrorBoundary");

var _DiffInspectWrapper = require("./DiffInspectWrapper");

var _DocumentChangeContext = require("./DocumentChangeContext");

var _FallbackDiff = require("./FallbackDiff");

var _RevertChangesButton = require("./RevertChangesButton");

var _ValueError = require("./ValueError");

var _FieldChange = _interopRequireDefault(require("./FieldChange.css"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function FieldChange(_ref) {
  var change = _ref.change;
  var DiffComponent = change.diffComponent || _FallbackDiff.FallbackDiff;

  var _useContext = (0, _react.useContext)(_DocumentChangeContext.DocumentChangeContext),
      documentId = _useContext.documentId,
      schemaType = _useContext.schemaType,
      rootDiff = _useContext.rootDiff,
      isComparingCurrent = _useContext.isComparingCurrent,
      _useContext$FieldWrap = _useContext.FieldWrapper,
      FieldWrapper = _useContext$FieldWrap === void 0 ? _react.default.Fragment : _useContext$FieldWrap;

  var docOperations = (0, _reactHooks.useDocumentOperation)(documentId, schemaType.name);

  var _React$useState = _react.default.useState(false),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      confirmRevertOpen = _React$useState2[0],
      setConfirmRevertOpen = _React$useState2[1];

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      revertHovered = _useState2[0],
      setRevertHovered = _useState2[1];

  var _useState3 = (0, _react.useState)(null),
      _useState4 = _slicedToArray(_useState3, 2),
      revertButtonElement = _useState4[0],
      setRevertButtonElement = _useState4[1];

  var handleRevertChanges = (0, _react.useCallback)(() => {
    (0, _undoChange.undoChange)(change, rootDiff, docOperations);
  }, [change, rootDiff, docOperations]);
  var handleRevertChangesConfirm = (0, _react.useCallback)(() => {
    setConfirmRevertOpen(true);
  }, []);

  var closeRevertChangesConfirmDialog = _react.default.useCallback(() => {
    setConfirmRevertOpen(false);
  }, []);

  var handleConfirmDialogAction = (0, _react.useCallback)(action => {
    if (action.action) action.action();
  }, []);
  var rootClass = (0, _classnames.default)(change.error ? _FieldChange.default.error : _FieldChange.default.root, revertHovered && _FieldChange.default.revertHovered);
  var handleRevertButtonMouseEnter = (0, _react.useCallback)(() => {
    setRevertHovered(true);
  }, []);
  var handleRevertButtonMouseLeave = (0, _react.useCallback)(() => {
    setRevertHovered(false);
  }, []);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: rootClass
  }, change.showHeader && /*#__PURE__*/_react.default.createElement("div", {
    className: _FieldChange.default.header
  }, /*#__PURE__*/_react.default.createElement(_ChangeBreadcrumb.ChangeBreadcrumb, {
    change: change,
    titlePath: change.titlePath
  })), /*#__PURE__*/_react.default.createElement(FieldWrapper, {
    path: change.path,
    hasHover: revertHovered
  }, /*#__PURE__*/_react.default.createElement(_DiffInspectWrapper.DiffInspectWrapper, {
    change: change,
    className: _FieldChange.default.change
  }, change.error ? /*#__PURE__*/_react.default.createElement(_ValueError.ValueError, {
    error: change.error
  }) : /*#__PURE__*/_react.default.createElement(_DiffErrorBoundary.DiffErrorBoundary, null, /*#__PURE__*/_react.default.createElement(_DiffContext.DiffContext.Provider, {
    value: {
      path: change.path
    }
  }, /*#__PURE__*/_react.default.createElement(DiffComponent, {
    diff: change.diff,
    schemaType: change.schemaType
  }))), isComparingCurrent && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: _FieldChange.default.revertChangesButtonContainer
  }, /*#__PURE__*/_react.default.createElement(_RevertChangesButton.RevertChangesButton, {
    onClick: handleRevertChangesConfirm,
    onMouseEnter: handleRevertButtonMouseEnter,
    onMouseLeave: handleRevertButtonMouseLeave,
    ref: setRevertButtonElement,
    selected: confirmRevertOpen
  })), confirmRevertOpen && /*#__PURE__*/_react.default.createElement(_popover.default, {
    portal: true,
    actions: [{
      color: 'danger',
      action: handleRevertChanges,
      title: 'Revert change'
    }, {
      kind: 'simple',
      action: closeRevertChangesConfirmDialog,
      title: 'Cancel'
    }],
    onAction: handleConfirmDialogAction // portal
    ,
    referenceElement: revertButtonElement,
    size: "small"
  }, "Are you sure you want to revert the changes?")))));
}
