// @sanity/field/lib/types/boolean/diff/BooleanFieldDiff.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BooleanFieldDiff = void 0;

var _react = _interopRequireDefault(require("react"));

var _diff = require("../../../diff");

var _preview = require("../preview");

var _BooleanFieldDiff = _interopRequireDefault(require("./BooleanFieldDiff.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BooleanFieldDiff = (_ref) => {
  var diff = _ref.diff,
      schemaType = _ref.schemaType;
  var fromValue = diff.fromValue,
      toValue = diff.toValue;
  var title = schemaType.title,
      options = schemaType.options;
  var Preview = (options === null || options === void 0 ? void 0 : options.layout) === 'checkbox' ? _preview.Checkbox : _preview.Switch;
  var userColor = (0, _diff.useDiffAnnotationColor)(diff, []) || {
    background: '',
    text: '',
    border: ''
  };
  var showToValue = toValue !== undefined && toValue !== null;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _BooleanFieldDiff.default.root
  }, /*#__PURE__*/_react.default.createElement(_diff.DiffTooltip, {
    diff: diff
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _BooleanFieldDiff.default.preview
  }, /*#__PURE__*/_react.default.createElement(Preview, {
    checked: fromValue,
    color: userColor
  }), showToValue && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_diff.FromToArrow, {
    style: {
      padding: '0 2px'
    }
  }), /*#__PURE__*/_react.default.createElement(Preview, {
    checked: toValue,
    color: userColor
  })))), showToValue && /*#__PURE__*/_react.default.createElement("div", {
    className: _BooleanFieldDiff.default.label
  }, title && /*#__PURE__*/_react.default.createElement("div", {
    className: _BooleanFieldDiff.default.title
  }, title)));
};

exports.BooleanFieldDiff = BooleanFieldDiff;
