// @sanity/components/lib/previews/CreateDocumentPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _router = require("part:@sanity/base/router");

var _react = _interopRequireDefault(require("react"));

var _CreateDocumentPreview = _interopRequireDefault(require("./CreateDocumentPreview.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULT_MEDIA_DIMENSION = {
  width: 80,
  height: 80,
  aspect: 1,
  fit: 'crop'
};

class CreateDocumentPreview extends _react.default.PureComponent {
  // eslint-disable-next-line complexity
  render() {
    var _this$props = this.props,
        _this$props$title = _this$props.title,
        title = _this$props$title === void 0 ? 'Untitled' : _this$props$title,
        subtitle = _this$props.subtitle,
        _this$props$media = _this$props.media,
        media = _this$props$media === void 0 ? this.props.icon : _this$props$media,
        isPlaceholder = _this$props.isPlaceholder,
        _this$props$mediaDime = _this$props.mediaDimensions,
        mediaDimensions = _this$props$mediaDime === void 0 ? DEFAULT_MEDIA_DIMENSION : _this$props$mediaDime,
        description = _this$props.description,
        params = _this$props.params,
        templateParams = _this$props.templateParams;

    if (isPlaceholder || !params) {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: _CreateDocumentPreview.default.placeholder
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: _CreateDocumentPreview.default.heading
      }, /*#__PURE__*/_react.default.createElement("h2", {
        className: _CreateDocumentPreview.default.title
      }, "Loading\u2026"), /*#__PURE__*/_react.default.createElement("h3", {
        className: _CreateDocumentPreview.default.subtitle
      }, "Loading\u2026")), media !== false && /*#__PURE__*/_react.default.createElement("div", {
        className: _CreateDocumentPreview.default.media
      }));
    }

    return /*#__PURE__*/_react.default.createElement(_router.IntentLink, {
      intent: "create",
      params: [params, templateParams],
      className: _CreateDocumentPreview.default.root,
      title: subtitle ? "Create new ".concat(title, " (").concat(subtitle, ")") : "Create new ".concat(title),
      onClick: this.props.onClick
    }, media !== false && /*#__PURE__*/_react.default.createElement("div", {
      className: _CreateDocumentPreview.default.media
    }, typeof media === 'function' && media({
      dimensions: mediaDimensions,
      layout: 'default'
    }), typeof media === 'string' && /*#__PURE__*/_react.default.createElement("div", {
      className: _CreateDocumentPreview.default.mediaString
    }, media), /*#__PURE__*/_react.default.isValidElement(media) && media), /*#__PURE__*/_react.default.createElement("div", {
      className: _CreateDocumentPreview.default.heading
    }, /*#__PURE__*/_react.default.createElement("h2", {
      className: _CreateDocumentPreview.default.title
    }, typeof title !== 'function' && title, typeof title === 'function' && title({
      layout: 'default'
    })), subtitle && /*#__PURE__*/_react.default.createElement("h3", {
      className: _CreateDocumentPreview.default.subtitle
    }, typeof subtitle === 'function' && subtitle({
      layout: 'default'
    }) || subtitle)), description && /*#__PURE__*/_react.default.createElement("p", {
      className: _CreateDocumentPreview.default.description
    }, description));
  }

}

var _default = CreateDocumentPreview;
exports.default = _default;
