// @sanity/components/lib/dialogs/stories/nested.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NestedStory = NestedStory;

var _dropdown = _interopRequireDefault(require("part:@sanity/components/buttons/dropdown"));

var _default = _interopRequireDefault(require("part:@sanity/components/dialogs/default"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _popover = _interopRequireDefault(require("part:@sanity/components/dialogs/popover"));

var _layer = require("part:@sanity/components/layer");

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function NestedStory() {
  return /*#__PURE__*/_react.default.createElement(_layer.LayerProvider, null, /*#__PURE__*/_react.default.createElement(DialogExample, null));
}

function DialogExample() {
  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      open2 = _useState2[0],
      setOpen2 = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      open3 = _useState4[0],
      setOpen3 = _useState4[1];

  var _useState5 = (0, _react.useState)(false),
      _useState6 = _slicedToArray(_useState5, 2),
      open4 = _useState6[0],
      setOpen4 = _useState6[1];

  var _useState7 = (0, _react.useState)(false),
      _useState8 = _slicedToArray(_useState7, 2),
      open5 = _useState8[0],
      setOpen5 = _useState8[1];

  var handleAction = (0, _react.useCallback)(action => {
    if (action.action) action.action();
  }, []);

  var _useState9 = (0, _react.useState)(null),
      _useState10 = _slicedToArray(_useState9, 2),
      refElement = _useState10[0],
      setRefElement = _useState10[1];

  return /*#__PURE__*/_react.default.createElement(_default.default, {
    actions: [{
      title: 'Open Dialog 2',
      action: () => setOpen2(true)
    }],
    onAction: handleAction,
    size: "large",
    title: "Dialog 1"
  }, /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    items: [{
      title: 'Hello'
    }, {
      title: 'Hello'
    }, {
      title: 'Hello'
    }],
    onAction: () => undefined
  }, "Test"), open2 && /*#__PURE__*/_react.default.createElement(_default.default, {
    actions: [{
      title: 'Open Dialog 3',
      action: () => setOpen3(true)
    }],
    onAction: handleAction,
    onClose: () => setOpen2(false),
    onClickOutside: () => setOpen2(false),
    size: "medium",
    title: "Dialog 2"
  }, /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    items: [{
      title: 'Hello'
    }, {
      title: 'Hello'
    }, {
      title: 'Hello'
    }],
    onAction: () => undefined
  }, "Test"), open3 && /*#__PURE__*/_react.default.createElement(_default.default, {
    actions: [{
      title: 'Open Dialog 4',
      action: () => setOpen4(true)
    }],
    onAction: handleAction,
    onClickOutside: () => setOpen3(false),
    onClose: () => setOpen3(false),
    onEscape: () => setOpen3(false),
    size: "small",
    title: "Dialog 3"
  }, /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    items: [{
      title: 'Hello'
    }, {
      title: 'Hello'
    }, {
      title: 'Hello'
    }],
    onAction: () => undefined
  }, "Test"), open4 && /*#__PURE__*/_react.default.createElement(_fullscreen.default, {
    onClickOutside: () => setOpen4(false),
    onClose: () => setOpen4(false),
    onEscape: () => setOpen4(false),
    title: "Dialog 4"
  }, /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    items: [{
      title: 'Hello'
    }, {
      title: 'Hello'
    }, {
      title: 'Hello'
    }],
    onAction: () => undefined
  }, "Test"), /*#__PURE__*/_react.default.createElement("button", {
    onClick: () => setOpen5(true),
    ref: setRefElement,
    type: "button"
  }, "Open popover dialog"), open5 && /*#__PURE__*/_react.default.createElement(_popover.default, {
    onClickOutside: () => setOpen5(false),
    onClose: () => setOpen5(false),
    onEscape: () => setOpen5(false),
    portal: true,
    referenceElement: refElement
  }, /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    items: [{
      title: 'Hello'
    }, {
      title: 'Hello'
    }, {
      title: 'Hello'
    }],
    onAction: () => undefined
  }, "Test"))))));
}
