// @sanity/default-layout/lib/stories/searchFieldDesktop.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SearchFieldDesktopStory = SearchFieldDesktopStory;

var _react = _interopRequireDefault(require("react"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _SearchField = _interopRequireDefault(require("../navbar/search/SearchField"));

var _SearchResults = _interopRequireDefault(require("../navbar/search/SearchResults"));

var _Navbar = _interopRequireDefault(require("../navbar/Navbar.css"));

var _DefaultLayout = _interopRequireDefault(require("../DefaultLayout.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function SearchFieldDesktopStory() {
  var hasResults = (0, _knobs.boolean)('hasResults', false, 'props');
  var items = hasResults ? [{
    hit: {
      _id: 'foo',
      _type: 'foo'
    },
    score: 12,
    stories: {
      path: 'foo',
      score: 1,
      why: 'test'
    }
  }] : [];
  var query = (0, _knobs.text)('query', '', 'props');
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
    className: _DefaultLayout.default.navBar
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.root
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _Navbar.default.search
  }, /*#__PURE__*/_react.default.createElement(_SearchField.default, {
    hotkeys: ['F'],
    isFocused: (0, _knobs.boolean)('isFocused', false, 'props'),
    isOpen: (0, _knobs.boolean)('isOpen', false, 'props'),
    results: /*#__PURE__*/_react.default.createElement(_SearchResults.default, {
      activeIndex: (0, _knobs.number)('activeIndex', -1, 'props'),
      isBleeding: false,
      isLoading: (0, _knobs.boolean)('isLoading', false, 'props'),
      items: items,
      query: query,
      renderItem: key => /*#__PURE__*/_react.default.createElement("div", {
        key: key.hit._id,
        style: {
          padding: '0.5em 0.75em'
        }
      }, key)
    }),
    value: query,
    onChange: () => (0, _actions.action)('onChange')
  })))));
}
