// @sanity/default-layout/lib/stories/toolSwitcher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToolMenuStory = ToolMenuStory;

var _pluginIcon = _interopRequireDefault(require("part:@sanity/base/plugin-icon"));

var _viewColumnIcon = _interopRequireDefault(require("part:@sanity/base/view-column-icon"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

var _ToolMenu = _interopRequireDefault(require("../navbar/toolMenu/ToolMenu"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ToolMenuStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
    defaultTone: "navbar"
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      outline: '1px solid red'
    }
  }, /*#__PURE__*/_react.default.createElement(_ToolMenu.default, {
    activeToolName: "desk",
    direction: "horizontal",
    isVisible: true,
    onSwitchTool: () => (0, _actions.action)('onSwitchTool'),
    router: {},
    tone: "navbar",
    tools: [{
      name: 'desk',
      title: 'Desk',
      icon: _viewColumnIcon.default
    }, {
      name: 'plugin1',
      title: 'Plugin 1',
      icon: _pluginIcon.default
    }, {
      name: 'plugin2',
      title: 'Plugin 2',
      icon: _pluginIcon.default
    }]
  })));
}
