// @sanity/components/lib/buttons/stories/anchor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnchorStory = AnchorStory;

var _sanityLogoIcon = _interopRequireDefault(require("part:@sanity/base/sanity-logo-icon"));

var _anchor = _interopRequireDefault(require("part:@sanity/components/buttons/anchor"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var actionsProps = {
  onBlur: (0, _actions.action)('onBlur'),
  onClick: (0, _actions.action)('onClick'),
  onFocus: (0, _actions.action)('onFocus')
};

function AnchorStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/buttons/anchor",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement(_anchor.default, _extends({}, actionsProps, {
    kind: (0, _knobs.select)('kind', ['default', 'simple', 'secondary'], 'default', 'props'),
    disabled: (0, _knobs.boolean)('disabled', false, 'props'),
    inverted: (0, _knobs.boolean)('inverted', false, 'props'),
    color: (0, _knobs.select)('color', [undefined, 'primary', 'success', 'danger', 'white'], undefined, 'props'),
    loading: (0, _knobs.boolean)('loading', false, 'props'),
    icon: (0, _knobs.boolean)('icon', false, 'props') ? _sanityLogoIcon.default : undefined,
    href: (0, _knobs.text)('href', 'http://example.org', 'props')
  }), (0, _knobs.text)('children', 'Touch Me!', 'props'))));
}
