// @sanity/components/lib/buttons/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _usersIcon = _interopRequireDefault(require("part:@sanity/base/users-icon"));

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var actionsProps = {
  onBlur: (0, _actions.action)('onBlur'),
  onClick: (0, _actions.action)('onClick'),
  onFocus: (0, _actions.action)('onFocus')
};
var buttonElement = null;

function testFocus() {
  if (buttonElement) buttonElement.focus();
}

function setButtonElement(element) {
  buttonElement = element;
}

function DefaultStory() {
  (0, _knobs.button)('Test focus', testFocus, 'test');

  var buttonProps = _objectSpread(_objectSpread({}, actionsProps), {}, {
    disabled: (0, _knobs.boolean)('Disabled', false, 'props'),
    inverted: (0, _knobs.boolean)('Inverted', false, 'props'),
    color: (0, _knobs.select)('Color', [undefined, 'primary', 'success', 'warning', 'danger', 'white'], undefined, 'props'),
    loading: (0, _knobs.boolean)('Loading', false, 'props'),
    icon: (0, _knobs.boolean)('Icon', undefined, 'props') ? _usersIcon.default : undefined,
    iconStatus: (0, _knobs.select)('Icon status', [undefined, 'primary', 'success', 'warning', 'danger'], undefined, 'props'),
    padding: (0, _knobs.select)('Padding', ['small', 'medium', 'large', 'none'], 'medium', 'props'),
    bleed: (0, _knobs.boolean)('Bleed', false, 'props'),
    selected: (0, _knobs.boolean)('Selected', false, 'props'),
    size: (0, _knobs.select)('Size', ['', 'extra-small', 'small', 'medium', 'large', 'extra-large'], '', 'props') || undefined,
    tone: (0, _knobs.select)('Tone', {
      '': '(none)',
      navbar: 'Navbar'
    }, '', 'Props') || undefined
  });

  var children = (0, _knobs.text)('Text', 'Label', 'props');
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/buttons/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 300
    }
  }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("h2", null, "kind=\"default\""), /*#__PURE__*/_react.default.createElement(_default.default, _extends({}, buttonProps, {
    ref: setButtonElement
  }), children)), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("h2", null, "kind=\"simple\""), /*#__PURE__*/_react.default.createElement(_default.default, _extends({}, buttonProps, {
    kind: "simple"
  }), children)))));
}
