// @sanity/components/lib/buttons/stories/dropDown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DropDownStory = DropDownStory;

var _sanityLogoIcon = _interopRequireDefault(require("part:@sanity/base/sanity-logo-icon"));

var _dropdown = _interopRequireDefault(require("part:@sanity/components/buttons/dropdown"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var items = [{
  index: '1',
  title: 'Test'
}, {
  index: '2',
  title: 'Test 2'
}, {
  index: '3',
  title: 'Test 3'
}, {
  index: '4',
  title: 'Test 4'
}, {
  index: '5',
  title: 'Test 5'
}, {
  index: '6',
  title: 'Test 6'
}, {
  index: '7',
  title: 'Test 7'
}, {
  index: '8',
  title: 'Test 8'
}, {
  index: '9',
  title: 'Test 9'
}, {
  index: '10',
  title: 'Test 10'
}, {
  index: '11',
  title: 'Test 11'
}];
var actionsProps = {
  onAction: (0, _actions.action)('onAction'),
  onBlur: (0, _actions.action)('onBlur'),
  onClick: (0, _actions.action)('onClick'),
  onFocus: (0, _actions.action)('onFocus')
};

function DropDownStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/buttons/dropdown",
    propTables: [_dropdown.default]
  }, /*#__PURE__*/_react.default.createElement(_dropdown.default, _extends({}, actionsProps, {
    items: (0, _knobs.object)('Items', items, 'props'),
    color: (0, _knobs.select)('Color', [undefined, 'primary', 'success', 'danger', 'white'], undefined, 'props'),
    kind: (0, _knobs.select)('Kind', ['default', 'simple', 'secondary'], 'default', 'props'),
    disabled: (0, _knobs.boolean)('Disabled', false, 'props'),
    inverted: (0, _knobs.boolean)('Inverted', false, 'props'),
    showArrow: (0, _knobs.boolean)('Show arrow', true, 'props'),
    loading: (0, _knobs.boolean)('Loading', false, 'props'),
    icon: (0, _knobs.boolean)('Icon', false, 'props') ? _sanityLogoIcon.default : undefined
  }), (0, _knobs.text)('Children', 'This is a dropdown', 'props'))));
}
