// @sanity/components/lib/buttons/stories/examples.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExamplesStory = ExamplesStory;

var _composeIcon = _interopRequireDefault(require("part:@sanity/base/compose-icon"));

var _sanityLogoIcon = _interopRequireDefault(require("part:@sanity/base/sanity-logo-icon"));

var _buttonGrid = _interopRequireDefault(require("part:@sanity/components/buttons/button-grid"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _dropdown = _interopRequireDefault(require("part:@sanity/components/buttons/dropdown"));

var _intent = _interopRequireDefault(require("part:@sanity/components/buttons/intent"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var preventDefault = evt => evt.preventDefault();

var items = [{
  index: '1',
  title: 'Test'
}, {
  index: '2',
  title: 'Test 2'
}, {
  index: '3',
  title: 'Test 3'
}, {
  index: '4',
  title: 'Test 4'
}, {
  index: '5',
  title: 'Test 5'
}, {
  index: '6',
  title: 'Test 6'
}, {
  index: '7',
  title: 'Test 7'
}, {
  index: '8',
  title: 'Test 8'
}, {
  index: '9',
  title: 'Test 9'
}, {
  index: '10',
  title: 'Test 10'
}, {
  index: '11',
  title: 'Test 11'
}];

function ExamplesStory() {
  var disabled = (0, _knobs.boolean)('Disabled', false);
  var loading = (0, _knobs.boolean)('Loading', false);
  return /*#__PURE__*/_react.default.createElement(_components.DebugRouterProvider, null, /*#__PURE__*/_react.default.createElement(_components.Container, {
    as: "form"
  }, /*#__PURE__*/_react.default.createElement("h2", null, "Default"), "Some text", /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('clicked'),
    disabled: disabled,
    loading: loading
  }, "Default"), "between", /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    disabled: disabled,
    loading: loading
  }, "Default with icon"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    inverted: true,
    disabled: disabled,
    loading: loading
  }, "Inverted"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    kind: "simple",
    disabled: disabled,
    loading: loading
  }, "Simple"), /*#__PURE__*/_react.default.createElement("h2", null, "Colors"), /*#__PURE__*/_react.default.createElement(_buttonGrid.default, null, /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    disabled: disabled,
    loading: loading
  }, "Undefined"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    color: "primary",
    disabled: disabled,
    loading: loading
  }, "Primary"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    color: "danger",
    disabled: disabled,
    loading: loading
  }, "Danger"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    color: "success",
    disabled: disabled,
    loading: loading
  }, "Success")), /*#__PURE__*/_react.default.createElement("h2", null, "Colors (inverted)"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    inverted: true,
    disabled: disabled,
    loading: loading
  }, "Undefined"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    color: "primary",
    inverted: true,
    disabled: disabled,
    loading: loading
  }, "Primary"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    color: "danger",
    inverted: true,
    disabled: disabled,
    loading: loading
  }, "Danger"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    color: "success",
    inverted: true,
    disabled: disabled,
    loading: loading
  }, "Success"), /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    items: items,
    onAction: (0, _actions.action)('onClick'),
    disabled: disabled,
    loading: loading,
    color: "primary",
    inverted: true
  }, "Dropdown"), /*#__PURE__*/_react.default.createElement("h2", null, "Colors (simple)"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    kind: "simple",
    disabled: disabled,
    loading: loading
  }, "Undefined"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    kind: "simple",
    color: "primary",
    disabled: disabled,
    loading: loading
  }, "Primary"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    kind: "simple",
    color: "danger",
    disabled: disabled,
    loading: loading
  }, "Danger"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    kind: "simple",
    color: "success",
    disabled: disabled,
    loading: loading
  }, "Success"), /*#__PURE__*/_react.default.createElement("h2", null, "With icons"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    disabled: disabled,
    loading: loading
  }, "With icon"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    color: "danger",
    icon: _sanityLogoIcon.default,
    disabled: disabled,
    loading: loading
  }, "Colored with icon"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    color: "danger",
    icon: _sanityLogoIcon.default,
    inverted: true,
    disabled: disabled,
    loading: loading
  }, "Danger, inverted & icon"), /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    icon: _sanityLogoIcon.default,
    inverted: true,
    color: "danger",
    items: items,
    onAction: (0, _actions.action)('onAction'),
    disabled: disabled,
    loading: loading
  }, "Dropdown"), /*#__PURE__*/_react.default.createElement("h2", null, "Only icons"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    title: "Default",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    color: "danger",
    title: "Danger",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    inverted: true,
    title: "Inverted",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    inverted: true,
    color: "danger",
    title: "Inverted danger",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    kind: "simple",
    title: "Simple",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    kind: "simple",
    color: "danger",
    title: "Simple danger",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    icon: _sanityLogoIcon.default,
    kind: "simple",
    color: "primary",
    title: "Simple primary",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement("h2", null, "Intent buttons with only icons"), /*#__PURE__*/_react.default.createElement(_intent.default, {
    intent: "create",
    params: {
      type: 'book'
    },
    onClick: preventDefault,
    icon: _composeIcon.default,
    title: "Default",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_intent.default, {
    intent: "create",
    params: {
      type: 'book'
    },
    onClick: preventDefault,
    icon: _composeIcon.default,
    color: "danger",
    title: "Danger",
    disabled: disabled
  }), /*#__PURE__*/_react.default.createElement(_intent.default, {
    intent: "create",
    params: {
      type: 'book'
    },
    onClick: preventDefault,
    icon: _composeIcon.default,
    inverted: true,
    title: "Inverted",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_intent.default, {
    intent: "create",
    params: {
      type: 'book'
    },
    onClick: preventDefault,
    icon: _composeIcon.default,
    inverted: true,
    color: "danger",
    title: "Inverted danger",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_intent.default, {
    intent: "create",
    params: {
      type: 'book'
    },
    onClick: preventDefault,
    icon: _composeIcon.default,
    kind: "simple",
    title: "Simple",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_intent.default, {
    intent: "create",
    params: {
      type: 'book'
    },
    onClick: preventDefault,
    icon: _composeIcon.default,
    kind: "simple",
    color: "danger",
    title: "Simple danger",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement(_intent.default, {
    intent: "create",
    params: {
      type: 'book'
    },
    onClick: preventDefault,
    icon: _composeIcon.default,
    kind: "simple",
    color: "primary",
    title: "Simple primary",
    disabled: disabled,
    loading: loading
  }), /*#__PURE__*/_react.default.createElement("h2", null, "On color areas"), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      backgroundColor: 'red',
      padding: '1rem'
    }
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    color: "white",
    disabled: disabled,
    loading: loading
  }, "White"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    kind: "simple",
    color: "white",
    disabled: disabled,
    loading: loading
  }, "White simple"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    inverted: true,
    disabled: disabled,
    loading: loading
  }, "Inverted"), /*#__PURE__*/_react.default.createElement(_default.default, {
    onClick: (0, _actions.action)('onClick'),
    inverted: true,
    color: "white",
    disabled: disabled,
    loading: loading
  }, "White inverted"))));
}
