// @sanity/components/lib/buttons/stories/grid.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridStory = GridStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _chance = _interopRequireDefault(require("chance"));

var _buttonGrid = _interopRequireDefault(require("part:@sanity/components/buttons/button-grid"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

function GridStory() {
  var qtyButtons = (0, _knobs.number)('# buttons', 2, 'test');
  var buttonText = (0, _knobs.text)('Button text', '');
  return /*#__PURE__*/_react.default.createElement(_components.Container, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/buttons/button-grid",
    propTables: [_buttonGrid.default]
  }, /*#__PURE__*/_react.default.createElement(_buttonGrid.default, {
    align: (0, _knobs.select)('align', ['start', 'end'], 'start', 'props'),
    secondary: (0, _range2.default)(0, (0, _knobs.number)('# secondary', 1)).map(i => {
      return /*#__PURE__*/_react.default.createElement(_default.default, {
        inverted: true,
        key: i
      }, buttonText || (i % 2 ? chance.word() : chance.name()));
    })
  }, (0, _range2.default)(0, qtyButtons).map(i => {
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      key: i
    }, buttonText || (i % 2 ? chance.word() : chance.name()));
  })), /*#__PURE__*/_react.default.createElement("h2", null, "Button grid with 1 (primary) button"), /*#__PURE__*/_react.default.createElement(_buttonGrid.default, {
    align: (0, _knobs.select)('align', ['start', 'end'], 'start', 'props')
  }, /*#__PURE__*/_react.default.createElement(_default.default, null, "Test")), /*#__PURE__*/_react.default.createElement("h2", null, "Button grid with 1 (secondary) button"), /*#__PURE__*/_react.default.createElement(_buttonGrid.default, {
    align: (0, _knobs.select)('align', ['start', 'end'], 'start', 'props'),
    secondary: /*#__PURE__*/_react.default.createElement(_default.default, null, "Test")
  })));
}
