// @sanity/components/lib/buttons/stories/group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GroupStory = GroupStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _chance = _interopRequireDefault(require("chance"));

var _sanityLogoIcon = _interopRequireDefault(require("part:@sanity/base/sanity-logo-icon"));

var _buttonGroup = _interopRequireDefault(require("part:@sanity/components/buttons/button-group"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _dropdown = _interopRequireDefault(require("part:@sanity/components/buttons/dropdown"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();
var items = [{
  index: '1',
  title: 'Test'
}, {
  index: '2',
  title: 'Test 2'
}, {
  index: '3',
  title: 'Test 3'
}, {
  index: '4',
  title: 'Test 4'
}, {
  index: '5',
  title: 'Test 5'
}, {
  index: '6',
  title: 'Test 6'
}, {
  index: '7',
  title: 'Test 7'
}, {
  index: '8',
  title: 'Test 8'
}, {
  index: '9',
  title: 'Test 9'
}, {
  index: '10',
  title: 'Test 10'
}, {
  index: '11',
  title: 'Test 11'
}];

function GroupStory() {
  var qtyButtons = (0, _knobs.number)('# buttons', 2, 'test');
  var loading = (0, _knobs.boolean)('Loading', false, 'buttonProp');
  var onlyIcon = (0, _knobs.boolean)('Only icon', false, 'test');
  var icon = (0, _knobs.boolean)('icon', false, 'buttonProp') ? _sanityLogoIcon.default : undefined;
  var buttonText = (0, _knobs.text)('Button text', '', 'buttonProp');
  var buttonColor = (0, _knobs.select)('kind', ['default', 'simple', 'secondary'], 'default', 'props');
  var buttonKind = (0, _knobs.select)('color', [undefined, 'primary', 'success', 'danger', 'white'], undefined, 'props');
  return /*#__PURE__*/_react.default.createElement(_components.Container, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/buttons/button-group",
    propTables: [_buttonGroup.default]
  }, /*#__PURE__*/_react.default.createElement(_buttonGroup.default, null, (0, _range2.default)(0, qtyButtons).map(i => {
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      kind: buttonKind,
      color: buttonColor,
      key: i,
      loading: loading,
      icon: icon || onlyIcon && _sanityLogoIcon.default
    }, buttonText || !onlyIcon && (i % 2 ? chance.word() : chance.name()));
  }), /*#__PURE__*/_react.default.createElement(_dropdown.default, {
    items: items,
    kind: buttonKind,
    color: buttonColor,
    loading: loading,
    onAction: (0, _actions.action)('onAction'),
    onFocus: (0, _actions.action)('onFocus'),
    onBlur: (0, _actions.action)('onBlur'),
    icon: icon || onlyIcon && _sanityLogoIcon.default
  }, !onlyIcon && 'Dropdown'))));
}
