// @sanity/components/lib/history/stories/listItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListItemStory = ListItemStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _listItem = _interopRequireDefault(require("part:@sanity/components/history/list-item"));

var _chance = _interopRequireDefault(require("chance"));

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

var mockUser = () => {
  return {
    identity: chance.guid(),
    displayName: chance.name(),
    email: chance.email(),
    imageUrl: "https://placeimg.com/64/64/any?".concat(Math.random() * 1000)
  };
};

function ListItemStory() {
  var users = (0, _range2.default)((0, _knobs.number)('users', 1, 'test')).map(() => mockUser());
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/history/list-item",
    propTables: [_listItem.default]
  }, /*#__PURE__*/_react.default.createElement(_listItem.default, {
    status: (0, _knobs.select)('status', ['published', 'unpublished', 'edited', 'created', 'truncated'], 'edited'),
    title: (0, _knobs.text)('title', '23 may 2019'),
    users: users,
    isCurrentVersion: (0, _knobs.boolean)('isCurrentVersion', false, 'props'),
    isSelected: (0, _knobs.boolean)('isSelected', false, 'props')
  }, (0, _knobs.boolean)('children', false, 'test') && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_default.default, {
    inverted: true,
    color: "white",
    padding: "small"
  }, "Publish"), /*#__PURE__*/_react.default.createElement(_default.default, {
    kind: "simple",
    padding: "small"
  }, "Discard"))));
}
