// @sanity/components/lib/history/stories/list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListStory = ListStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _react = _interopRequireDefault(require("react"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _listItem = _interopRequireDefault(require("part:@sanity/components/history/list-item"));

var _chance = _interopRequireDefault(require("chance"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var chance = new _chance.default();

var mockUser = () => {
  return {
    identity: chance.guid(),
    displayName: chance.name(),
    email: chance.email(),
    imageUrl: "https://placeimg.com/64/64/any?".concat(Math.random() * 1000)
  };
};

var mockHistoryItem = () => {
  return {
    title: 'Test',
    status: 'published',
    key: chance.guid(),
    users: (0, _range2.default)(Math.floor(Math.random() * 4) + 1).map(() => mockUser())
  };
};

function ListStory() {
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_listItem.default, {
    status: "draft",
    title: "test",
    users: [mockUser()]
  }), /*#__PURE__*/_react.default.createElement(_listItem.default, {
    status: "published",
    title: "23 may 2018",
    users: [mockUser()],
    isSelected: true,
    isCurrentVersion: true
  }), /*#__PURE__*/_react.default.createElement(_listItem.default, {
    status: "unpublished",
    title: "22 may 2018",
    users: [mockUser()]
  }), (0, _range2.default)((0, _knobs.number)('items', 10, 'test')).map(() => {
    var item = mockHistoryItem();
    return /*#__PURE__*/_react.default.createElement(_listItem.default, _extends({}, item, {
      key: item.key
    }));
  }), /*#__PURE__*/_react.default.createElement(_listItem.default, {
    status: "created",
    title: "test",
    users: [mockUser()]
  }));
}
