// @sanity/components/lib/toggles/stories/buttons.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonsStory = ButtonsStory;

var _react = _interopRequireDefault(require("react"));

var _storybook = require("part:@sanity/storybook");

var _buttons = _interopRequireDefault(require("part:@sanity/components/toggles/buttons"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var centerStyle = {
  display: 'block',
  position: 'absolute',
  padding: '2rem',
  boxSizing: 'border-box',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)'
};

function ButtonsStory() {
  var items = [{
    title: 'The good',
    key: 'good'
  }, {
    title: 'The Bad',
    key: 'bad'
  }, {
    title: 'The ugly',
    key: 'ugly'
  }];
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/toggles/buttons",
    propTables: [_buttons.default]
  }, /*#__PURE__*/_react.default.createElement(_buttons.default, {
    items: items,
    label: "Select something",
    onChange: (0, _storybook.action)('onChange'),
    value: items[(0, _knobs.number)('value', 0, {
      range: 'true',
      min: 0,
      max: 2
    }, 'test')]
  })));
}
