// @sanity/components/lib/toggles/stories/checkbox.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CheckboxStory = CheckboxStory;

var _react = _interopRequireDefault(require("react"));

var _storybook = require("part:@sanity/storybook");

var _checkbox = _interopRequireDefault(require("part:@sanity/components/toggles/checkbox"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var centerStyle = {
  display: 'block',
  position: 'absolute',
  padding: '2rem',
  boxSizing: 'border-box',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)'
};

function CheckboxStory() {
  var fontSize = (0, _knobs.number)('fontSize', 1, {
    range: true,
    min: 0.5,
    max: 3,
    step: 0.1
  }, 'test');
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/toggles/checkbox",
    propTables: [_checkbox.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      fontSize: "".concat(fontSize, "em")
    }
  }, /*#__PURE__*/_react.default.createElement(_checkbox.default, {
    label: (0, _knobs.text)('label', "This is the label (".concat(fontSize, "em)"), 'props'),
    checked: (0, _knobs.boolean)('undefined', false) ? undefined : (0, _knobs.boolean)('checked', false, 'props'),
    disabled: (0, _knobs.boolean)('disabled', false, 'props'),
    onChange: (0, _storybook.action)('onChange'),
    onBlur: (0, _storybook.action)('onBlur'),
    onFocus: (0, _storybook.action)('onFocus'),
    ref: {
      current: true
    }
  }, (0, _knobs.boolean)('Children', false, 'test') ? /*#__PURE__*/_react.default.createElement("h1", {
    style: {
      color: 'red'
    }
  }, "Test") : false))));
}
