// @sanity/components/lib/toggles/stories/switch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SwitchStory = SwitchStory;

var _react = _interopRequireDefault(require("react"));

var _storybook = require("part:@sanity/storybook");

var _switch = _interopRequireDefault(require("part:@sanity/components/toggles/switch"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var centerStyle = {
  display: 'block',
  position: 'absolute',
  padding: '2rem',
  boxSizing: 'border-box',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)'
};

function SwitchStory() {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/toggles/switch",
    propTables: [_switch.default]
  }, /*#__PURE__*/_react.default.createElement(_switch.default, {
    checked: (0, _knobs.boolean)('undefined', false) ? undefined : (0, _knobs.boolean)('checked', false, 'props'),
    label: (0, _knobs.text)('label', 'This is the label', 'props'),
    description: (0, _knobs.text)('description', 'This is the description', 'props'),
    disabled: (0, _knobs.boolean)('disabled', false, 'props'),
    onChange: (0, _storybook.action)('change'),
    onFocus: (0, _storybook.action)('onFocus'),
    onBlur: (0, _storybook.action)('onBlur')
  })));
}
