// @sanity/components/lib/toggles/stories/toggleButtonCollectionStory.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToggleButtonCollectionStory = ToggleButtonCollectionStory;

var _react = _interopRequireDefault(require("react"));

var _storybook = require("part:@sanity/storybook");

var _button = _interopRequireDefault(require("part:@sanity/components/toggles/button"));

var _sanityLogoIcon = _interopRequireDefault(require("part:@sanity/base/sanity-logo-icon"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var centerStyle = {
  display: 'block',
  position: 'absolute',
  padding: '2rem',
  boxSizing: 'border-box',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)'
};

function ToggleButtonCollectionStory() {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    icon: _sanityLogoIcon.default,
    selected: true,
    onClick: (0, _storybook.action)('onClick')
  }), /*#__PURE__*/_react.default.createElement(_button.default, {
    icon: _sanityLogoIcon.default,
    selected: true,
    onClick: (0, _storybook.action)('onClick')
  }), /*#__PURE__*/_react.default.createElement(_button.default, {
    icon: _sanityLogoIcon.default,
    selected: true,
    onClick: (0, _storybook.action)('onClick')
  }), /*#__PURE__*/_react.default.createElement(_button.default, {
    icon: _sanityLogoIcon.default,
    selected: true,
    onClick: (0, _storybook.action)('onClick')
  }));
}
