// @sanity/components/lib/toggles/stories/toggleButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ToggleButtonStory = ToggleButtonStory;

var _react = _interopRequireDefault(require("react"));

var _storybook = require("part:@sanity/storybook");

var _buttons = _interopRequireDefault(require("part:@sanity/components/toggles/buttons"));

var _button = _interopRequireDefault(require("part:@sanity/components/toggles/button"));

var _sanityLogoIcon = _interopRequireDefault(require("part:@sanity/base/sanity-logo-icon"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var centerStyle = {
  display: 'block',
  position: 'absolute',
  padding: '2rem',
  boxSizing: 'border-box',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)'
};

function ToggleButtonStory() {
  var icon = (0, _knobs.boolean)('icon', false) ? _sanityLogoIcon.default : false;
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/toggles/buttons",
    propTables: [_buttons.default]
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    selected: (0, _knobs.boolean)('selected', false, 'props'),
    disabled: (0, _knobs.boolean)('disabled', false, 'props'),
    onClick: (0, _storybook.action)('onClick'),
    icon: icon || undefined
  }, (0, _knobs.text)('children', 'this is the content', 'props'))));
}
