// @sanity/components/lib/progress/stories/progressCircle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProgressCircleStory = ProgressCircleStory;

var _circle = _interopRequireDefault(require("part:@sanity/components/progress/circle"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var centerStyle = {
  position: 'absolute',
  top: '50%',
  left: '50%',
  transform: 'translate(-50%, -50%)',
  width: '100%',
  maxWidth: '600px',
  padding: '1rem'
};

function ProgressCircleStory() {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/progress/circle",
    propTables: [_circle.default]
  }, /*#__PURE__*/_react.default.createElement(_circle.default, {
    isComplete: (0, _knobs.boolean)('isComplete', false, 'props'),
    isInProgress: (0, _knobs.boolean)('isInProgress', false, 'props'),
    percent: (0, _knobs.number)('percent', 10, {
      range: true,
      min: 0,
      max: 100,
      step: 1
    }, 'props'),
    showPercent: (0, _knobs.boolean)('showPercent', true, 'props'),
    text: (0, _knobs.text)('text', 'Uploaded', 'props')
  })));
}
