// @sanity/components/lib/progress/stories/progressBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProgressBarStory = ProgressBarStory;

var _bar = _interopRequireDefault(require("part:@sanity/components/progress/bar"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ProgressBarStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 350
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/progress/bar",
    propTables: [_bar.default]
  }, /*#__PURE__*/_react.default.createElement(_bar.default, {
    isComplete: (0, _knobs.boolean)('isComplete', false, 'props'),
    isInProgress: (0, _knobs.boolean)('isInProgress', false, 'props'),
    percent: (0, _knobs.number)('percentage', 10, {
      range: true,
      min: 0,
      max: 100,
      step: 1
    }, 'props'),
    showPercent: (0, _knobs.boolean)('showPercent', false, 'props'),
    text: (0, _knobs.text)('text', 'Downloaded 5.1 of 8.2Mb', 'props')
  }))));
}
