// @sanity/desk-tool/lib/panes/documentPane/documentPanel/header/validationMenu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ValidationMenu = ValidationMenu;

var _menuButton = require("part:@sanity/components/menu-button");

var _list = _interopRequireDefault(require("part:@sanity/components/validation/list"));

var _errorOutlineIcon = _interopRequireDefault(require("part:@sanity/base/error-outline-icon"));

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ValidationMenu(props) {
  var boundaryElement = props.boundaryElement,
      isOpen = props.isOpen,
      markers = props.markers,
      schemaType = props.schemaType,
      setFocusPath = props.setFocusPath,
      setOpen = props.setOpen;
  var validationMarkers = markers.filter(marker => marker.type === 'validation');
  var validationErrorMarkers = validationMarkers.filter(marker => marker.level === 'error');
  var validationWarningwarnings = validationMarkers.filter(marker => marker.level === 'warning');
  var handleClose = (0, _react.useCallback)(() => setOpen(false), [setOpen]);

  if (validationErrorMarkers.length === 0 && validationWarningwarnings.length === 0) {
    return null;
  }

  var popoverContent = /*#__PURE__*/_react.default.createElement(_list.default, {
    documentType: schemaType,
    markers: validationMarkers,
    onClose: handleClose,
    onFocus: setFocusPath // showLink

  });

  return /*#__PURE__*/_react.default.createElement(_menuButton.MenuButton, {
    boundaryElement: boundaryElement,
    buttonProps: {
      color: validationErrorMarkers.length > 0 ? 'danger' : 'warning',
      kind: 'simple',
      icon: _errorOutlineIcon.default,
      padding: 'small',
      selected: isOpen,
      title: 'Show validation issues'
    },
    menu: popoverContent,
    open: isOpen,
    placement: "bottom",
    setOpen: setOpen
  });
}
