// @sanity/core/lib/commands/hook/listHooksCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var _default = {
  name: 'list',
  group: 'hook',
  signature: '',
  description: 'List hooks for a given project',
  action: function () {
    var _action = _asyncToGenerator(function* (args, context) {
      var apiClient = context.apiClient,
          output = context.output;
      var client = apiClient();
      var hooks;

      try {
        hooks = yield client.request({
          uri: '/hooks'
        });
      } catch (err) {
        throw new Error("Hook list retrieval failed:\n".concat(err.message));
      }

      hooks.forEach(hook => {
        output.print("Name: ".concat(hook.name));
        output.print("Dataset: ".concat(hook.dataset));
        output.print("URL: ".concat(hook.url));
        output.print('');
      });
    });

    function action(_x, _x2) {
      return _action.apply(this, arguments);
    }

    return action;
  }()
};
exports.default = _default;
