// sanity-plugin-asset-source-unsplash/dist/lib/datastores/unsplash.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var client_1 = __importDefault(require("part:@sanity/base/client"));
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var fetchSearch = function (query, page, perPage) {
    return rxjs_1.defer(function () {
        return client_1.default.observable.request({
            url: "/addons/unsplash/search/photos?query=" + encodeURIComponent(query) + "&page=" + page + "&per_page=" + perPage,
            withCredentials: true,
            method: 'GET'
        });
    });
};
var fetchList = function (type, page, perPage) {
    return rxjs_1.defer(function () {
        return client_1.default.observable.request({
            url: "/addons/unsplash/photos?order_by=" + type + "&page=" + page + "&per_page=" + perPage,
            withCredentials: true,
            method: 'GET'
        });
    });
};
function fetchDownloadUrl(photo) {
    var downloadUrl = photo.links.download_location.replace('https://api.unsplash.com', '/addons/unsplash');
    return client_1.default
        .request({
        url: downloadUrl,
        withCredentials: true,
        method: 'GET'
    })
        .then(function (result) {
        return result.url;
    });
}
exports.fetchDownloadUrl = fetchDownloadUrl;
exports.sanityClient = client_1.default;
exports.search = function (query, page, resultsPerPage) {
    return rxjs_1.concat(query.pipe(operators_1.withLatestFrom(page), operators_1.debounceTime(500), operators_1.distinctUntilChanged(), operators_1.switchMap(function (_a) {
        var q = _a[0], p = _a[1];
        if (q) {
            return fetchSearch(q, p, resultsPerPage).pipe(operators_1.distinctUntilChanged(), operators_1.map(function (result) { return result.results; }));
        }
        return fetchList('popular', p, resultsPerPage);
    })));
};
