// @sanity/form-builder/lib/inputs/DateInputs/DateInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _moment = _interopRequireDefault(require("moment"));

var _react = _interopRequireDefault(require("react"));

var _PatchEvent = _interopRequireWildcard(require("../../PatchEvent"));

var _BaseDateTimeInput = _interopRequireDefault(require("./BaseDateTimeInput"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// This is the format dates are stored on
var VALUE_FORMAT = 'YYYY-MM-DD'; // default to how they are stored

var DEFAULT_DATE_FORMAT = VALUE_FORMAT;

function parseOptions() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return {
    dateFormat: options.dateFormat || DEFAULT_DATE_FORMAT,
    calendarTodayLabel: options.calendarTodayLabel || 'Today'
  };
}

class DateInput extends _react.default.Component {
  constructor() {
    super(...arguments);

    _defineProperty(this, "baseDateTimeInputRef", null);

    _defineProperty(this, "handleChange", nextMoment => {
      var patch = nextMoment ? (0, _PatchEvent.set)(nextMoment.format(VALUE_FORMAT)) : (0, _PatchEvent.unset)();
      this.props.onChange(_PatchEvent.default.from([patch]));
    });

    _defineProperty(this, "setBaseInput", baseInput => {
      this.baseDateTimeInputRef = baseInput;
    });
  }

  focus() {
    if (this.baseDateTimeInputRef) {
      this.baseDateTimeInputRef.focus();
    }
  }

  render() {
    var _this$props = this.props,
        value = _this$props.value,
        markers = _this$props.markers,
        type = _this$props.type,
        readOnly = _this$props.readOnly,
        level = _this$props.level,
        onFocus = _this$props.onFocus,
        presence = _this$props.presence;
    var title = type.title,
        description = type.description;
    var momentValue = value ? (0, _moment.default)(value) : null;
    var options = parseOptions(type.options);
    return /*#__PURE__*/_react.default.createElement(_BaseDateTimeInput.default, {
      dateOnly: true,
      ref: this.setBaseInput,
      value: momentValue,
      readOnly: readOnly,
      level: level,
      title: title,
      description: description,
      placeholder: type.placeholder,
      markers: markers,
      dateFormat: options.dateFormat,
      todayLabel: options.calendarTodayLabel,
      onChange: this.handleChange,
      onFocus: onFocus,
      presence: presence
    });
  }

}

exports.default = DateInput;
