// @sanity/components/lib/lists/stories/cards.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardsStory = CardsStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _chance = _interopRequireDefault(require("chance"));

var _grid = require("part:@sanity/components/lists/grid");

var _card = _interopRequireDefault(require("part:@sanity/components/previews/card"));

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

function CardsStory() {
  var items = (0, _range2.default)(50).map((_, i) => {
    var width = 300;
    var height = 120;
    var randomImage = "http://placekitten.com/".concat(width, "/").concat(height);
    return {
      key: "".concat(i),
      title: chance.name(),
      subtitle: chance.sentence(),
      description: chance.sentence({
        words: 1
      }),
      media: /*#__PURE__*/_react.default.createElement("img", {
        src: randomImage,
        height: height,
        width: width
      })
    };
  });
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/lists/grid",
    propTables: [_grid.List]
  }, /*#__PURE__*/_react.default.createElement(_grid.List, null, items.map(item => /*#__PURE__*/_react.default.createElement(_grid.Item, {
    key: item.key
  }, /*#__PURE__*/_react.default.createElement(_card.default, item)))));
}
