// @sanity/components/lib/lists/stories/createDocument.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateDocumentStory = CreateDocumentStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _chance = _interopRequireDefault(require("chance"));

var _fileIcon = _interopRequireDefault(require("part:@sanity/base/file-icon"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _createDocument = _interopRequireDefault(require("part:@sanity/components/lists/create-document"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

function CreateDocumentStory() {
  var templateChoices = (0, _range2.default)((0, _knobs.number)('# Choices', 5, 'test')).map((_, i) => {
    return {
      id: "".concat(i),
      key: "".concat(i),
      title: chance.animal(),
      subtitle: 'test',
      template: 'test',
      icon: _fileIcon.default
    };
  });
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/lists/create-document",
    propTables: [_createDocument.default]
  }, /*#__PURE__*/_react.default.createElement(_createDocument.default, {
    items: templateChoices
  }));
}
