// @sanity/components/lib/lists/stories/mediaPreview.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MediaPreviewStory = MediaPreviewStory;

var _random2 = _interopRequireDefault(require("lodash/random"));

var _range2 = _interopRequireDefault(require("lodash/range"));

var _chance = _interopRequireDefault(require("chance"));

var _grid = require("part:@sanity/components/lists/grid");

var _media = _interopRequireDefault(require("part:@sanity/components/previews/media"));

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

function MediaPreviewStory() {
  var items = (0, _range2.default)(50).map((_, i) => {
    var width = (0, _random2.default)(10, 80) * 10;
    var height = (0, _random2.default)(10, 50) * 10;
    var randomImage = "http://placekitten.com/".concat(width, "/").concat(height);
    return {
      key: "".concat(i),
      title: chance.name(),
      imageUrl: randomImage,
      aspect: width / height
    };
  });
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/lists/grid",
    propTables: [_grid.List]
  }, /*#__PURE__*/_react.default.createElement(_grid.List, null, items.map(item => /*#__PURE__*/_react.default.createElement(_grid.Item, {
    key: item.key
  }, /*#__PURE__*/_react.default.createElement(_media.default, item)))));
}
