// @sanity/components/lib/utilities/stories/activateOnFocus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActivateOnFocusStory = ActivateOnFocusStory;

var _react = _interopRequireDefault(require("react"));

var _storybook = require("part:@sanity/storybook");

var _switch = _interopRequireDefault(require("part:@sanity/components/toggles/switch"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _activateOnFocus = _interopRequireDefault(require("part:@sanity/components/utilities/activate-on-focus"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ActivateOnFocusStory() {
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/utilities/activate-on-focus",
    propTables: [_switch.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      height: '700px',
      width: '500px',
      border: '3px dotted #ccc',
      position: 'relative',
      padding: '20px'
    }
  }, /*#__PURE__*/_react.default.createElement(_activateOnFocus.default, {
    onActivate: (0, _storybook.action)('onActivate') // onBlur={action('onBlur')}
    ,
    message: (0, _knobs.text)('message', '', 'props') // enableBlur={boolean('enableBlur', false, 'props')}

  }, /*#__PURE__*/_react.default.createElement("textarea", {
    rows: 30
  }, "This should not be selected on first click"))));
}
