// @sanity/components/lib/textareas/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _storybook = require("part:@sanity/storybook");

var _default = _interopRequireDefault(require("part:@sanity/components/textareas/default"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 640
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/textinputs/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    isClearable: (0, _knobs.boolean)('isClearable', false, 'props'),
    placeholder: (0, _knobs.text)('placeholder', 'This is the placeholder', 'props'),
    onChange: (0, _storybook.action)('onChange'),
    onClear: (0, _storybook.action)('onClear'),
    onFocus: (0, _storybook.action)('onFocus'),
    onKeyPress: (0, _storybook.action)('onKeyPress'),
    onBlur: (0, _storybook.action)('onBlur'),
    rows: (0, _knobs.number)('rows', 2, 'props'),
    value: (0, _knobs.text)('value', '', 'props'),
    disabled: (0, _knobs.boolean)('disabled', false, 'props'),
    readOnly: (0, _knobs.boolean)('readOnly', false, 'props')
  }))));
}
