// @sanity/components/lib/avatar/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _avatar = require("part:@sanity/components/avatar");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  var props = {
    animateArrowFrom: (0, _knobs.select)('Animate arrow from', {
      '': '(none)',
      top: 'Top',
      inside: 'Inside',
      bottom: 'Bottom'
    }, '', 'Props') || undefined,
    arrowPosition: (0, _knobs.select)('Arrow position', {
      '': '(none)',
      top: 'Top',
      inside: 'Inside',
      bottom: 'Bottom'
    }, '', 'Props') || undefined,
    color: {
      dark: (0, _knobs.color)('Color (dark)', '#0f0', 'Props'),
      light: (0, _knobs.color)('Color (light)', '#0c0', 'Props')
    },
    initials: (0, _knobs.text)('Intitials', 'si', 'Props'),
    size: (0, _knobs.select)('Size', {
      small: 'Small',
      medium: 'Medium',
      large: 'Large'
    }, 'small', 'Props'),
    src: (0, _knobs.text)('Image URL', undefined, 'Props'),
    status: (0, _knobs.select)('Status', {
      '': '(none)',
      online: 'Online',
      editing: 'Editing',
      inactive: 'Inactive'
    }, '', 'Props') || undefined,
    title: (0, _knobs.text)('Title', 'Sanity.io', 'Props'),
    tone: (0, _knobs.select)('Tone', {
      '': '(none)',
      navbar: 'Navbar'
    }, '', 'Props') || undefined
  };
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/dialogs/default",
    propTables: [_avatar.Avatar]
  }, /*#__PURE__*/_react.default.createElement(_avatar.Avatar, props)));
}
