// @sanity/base/lib/presence/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FormFieldPresenceContext", {
  enumerable: true,
  get: function get() {
    return _context.FormFieldPresenceContext;
  }
});
Object.defineProperty(exports, "FormFieldPresence", {
  enumerable: true,
  get: function get() {
    return _types.FormFieldPresence;
  }
});
Object.defineProperty(exports, "GlobalPresence", {
  enumerable: true,
  get: function get() {
    return _types2.GlobalPresence;
  }
});
Object.defineProperty(exports, "PresenceOverlay", {
  enumerable: true,
  get: function get() {
    return _PresenceOverlay.PresenceOverlay;
  }
});
Object.defineProperty(exports, "FieldPresence", {
  enumerable: true,
  get: function get() {
    return _FieldPresence.FieldPresence;
  }
});
Object.defineProperty(exports, "PresenceScope", {
  enumerable: true,
  get: function get() {
    return _PresenceScope.PresenceScope;
  }
});

var _context = require("./context");

var _types = require("./types");

var _types2 = require("../datastores/presence/types");

var _PresenceOverlay = require("./overlay/PresenceOverlay");

var _FieldPresence = require("./FieldPresence");

var _PresenceScope = require("./PresenceScope");
