// @sanity/base/lib/presence/PresenceTooltip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PresenceTooltip = PresenceTooltip;

var _tooltip = require("part:@sanity/components/tooltip");

var _react = _interopRequireDefault(require("react"));

var _UserAvatar = require("../components/UserAvatar");

var _PresenceTooltip = _interopRequireDefault(require("./PresenceTooltip.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/require-default-props */
function PresenceTooltip(props) {
  var children = props.children,
      items = props.items,
      placement = props.placement;

  var content = /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceTooltip.default.root
  }, items.map(item => /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceTooltip.default.item,
    key: item.user.id
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceTooltip.default.avatarContainer
  }, /*#__PURE__*/_react.default.createElement(_UserAvatar.UserAvatar, {
    user: item.user,
    status: "online"
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceTooltip.default.textContainer
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _PresenceTooltip.default.item__displayName
  }, item.user.displayName)))));

  return /*#__PURE__*/_react.default.createElement(_tooltip.Tooltip, {
    content: content,
    placement: placement
  }, children);
}
