// @sanity/form-builder/lib/inputs/TagsArrayInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

var _textfield = _interopRequireDefault(require("part:@sanity/components/tags/textfield"));

var _PatchEvent = _interopRequireWildcard(require("../../PatchEvent"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class TagsArrayInput extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_input", void 0);

    _defineProperty(this, "_inputId", (0, _uniqueId2.default)('TagsArrayInput'));

    _defineProperty(this, "handleChange", nextValue => {
      this.set(nextValue);
    });

    _defineProperty(this, "setInput", el => {
      this._input = el;
    });
  }

  set(nextValue) {
    var patch = nextValue.length === 0 ? (0, _PatchEvent.unset)() : (0, _PatchEvent.set)(nextValue);
    this.props.onChange(_PatchEvent.default.from(patch));
  }

  focus() {
    if (this._input) {
      this._input.focus();
    }
  }

  render() {
    var _this$props = this.props,
        type = _this$props.type,
        value = _this$props.value,
        readOnly = _this$props.readOnly,
        level = _this$props.level,
        onFocus = _this$props.onFocus,
        presence = _this$props.presence;
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      level: level,
      label: type.title,
      description: type.description,
      presence: presence,
      labelFor: this._inputId
    }, /*#__PURE__*/_react.default.createElement(_textfield.default, {
      inputId: this._inputId,
      readOnly: readOnly,
      value: value,
      onChange: this.handleChange,
      onFocus: onFocus,
      ref: this.setInput
    }));
  }

}

exports.default = TagsArrayInput;
