// @sanity/base/lib/preview/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _components.SanityPreview;
  }
});
Object.defineProperty(exports, "PreviewSubscriber", {
  enumerable: true,
  get: function get() {
    return _components.PreviewSubscriber;
  }
});
Object.defineProperty(exports, "PreviewFields", {
  enumerable: true,
  get: function get() {
    return _components.PreviewFields;
  }
});
Object.defineProperty(exports, "SanityDefaultPreview", {
  enumerable: true,
  get: function get() {
    return _components.SanityDefaultPreview;
  }
});
Object.defineProperty(exports, "observeForPreview", {
  enumerable: true,
  get: function get() {
    return _preview.observeForPreview;
  }
});
Object.defineProperty(exports, "observeWithPaths", {
  enumerable: true,
  get: function get() {
    return _preview.observeWithPaths;
  }
});
Object.defineProperty(exports, "materializePaths", {
  enumerable: true,
  get: function get() {
    return _preview.materializePaths;
  }
});
Object.defineProperty(exports, "observePaths", {
  enumerable: true,
  get: function get() {
    return _preview.observePaths;
  }
});

var _components = require("@sanity/preview/components");

var _preview = require("@sanity/preview");
