// sanity-plugin-asset-source-unsplash/dist/lib/components/Photo.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var Photo_css_1 = __importDefault(require("./Photo.css"));
var UTM_SOURCE = 'sanity-plugin-asset-source-unsplash';
var Photo = /** @class */ (function (_super) {
    __extends(Photo, _super);
    function Photo() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.rootElm = react_1.default.createRef();
        _this.handleClick = function (event) {
            _this.props.onClick(_this.props.data);
        };
        _this.handleCreditLineClicked = function (event) {
            event.stopPropagation();
            var data = _this.props.data;
            var url = data.links.html + "?utm_source=" + encodeURIComponent(UTM_SOURCE) + "&utm_medium=referral";
            window.open(url, data.id, 'noreferrer,noopener');
        };
        _this.handleKeyDown = function (event) {
            var _a = _this.props, onKeyDown = _a.onKeyDown, data = _a.data;
            onKeyDown(event);
            if (event.keyCode === 13) {
                _this.props.onClick(data);
            }
        };
        _this.handleMouseDown = function (event) {
            _this.props.onFocus(_this.props.data);
        };
        return _this;
    }
    Photo.prototype.componentDidUpdate = function (prevProps) {
        if (!prevProps.active && this.props.active && this.rootElm.current) {
            this.rootElm.current.focus();
            this.props.onFocus(this.props.data);
        }
    };
    Photo.prototype.render = function () {
        var _a = this.props, width = _a.width, height = _a.height, data = _a.data, active = _a.active;
        var src = data.urls.small;
        var userName = data.user.name;
        return (react_1.default.createElement("a", { ref: this.rootElm, href: "#", className: "" + Photo_css_1.default.root, title: "Select image by " + userName + " from Unsplash", tabIndex: 0, onKeyDown: this.handleKeyDown, onMouseDown: this.handleMouseDown, style: {
                width: width + "px",
                height: height + "px",
                backgroundImage: "url(\"" + src + "\")"
            }, onClick: this.handleClick },
            react_1.default.createElement("div", { className: Photo_css_1.default.creditLine, onClick: this.handleCreditLineClicked },
                react_1.default.createElement("div", { className: Photo_css_1.default.creditLineLink, title: "Open image by " + userName + " on Unsplash in new window" },
                    "By @",
                    data.user.username))));
    };
    return Photo;
}(react_1.default.Component));
exports.default = Photo;
