// @sanity/form-builder/lib/inputs/OptionsArrayInput/Item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _types = require("@sanity/types");

var _checkbox = _interopRequireDefault(require("part:@sanity/components/toggles/checkbox"));

var _Preview = _interopRequireDefault(require("../../Preview"));

var _legacyOptionsSupport = require("./legacyOptionsSupport");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Item extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "handleChange", event => {
      var _this$props = this.props,
          onChange = _this$props.onChange,
          value = _this$props.value;
      onChange(event.currentTarget.checked, value);
    });
  }

  render() {
    var _this$props2 = this.props,
        value = _this$props2.value,
        checked = _this$props2.checked,
        type = _this$props2.type,
        onFocus = _this$props2.onFocus,
        onBlur = _this$props2.onBlur,
        readOnly = _this$props2.readOnly;
    return /*#__PURE__*/_react.default.createElement(_checkbox.default, {
      onChange: this.handleChange,
      onFocus: onFocus,
      onBlur: onBlur,
      checked: checked,
      readOnly: readOnly || type.readOnly,
      label: (0, _types.isTitledListValue)(value) ? value.title : /*#__PURE__*/_react.default.createElement(_Preview.default, {
        layout: "inline",
        type: type,
        value: (0, _legacyOptionsSupport.resolveValueWithLegacyOptionsSupport)(value)
      })
    });
  }

}

exports.default = Item;
