// @sanity/form-builder/lib/inputs/common/Details.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _angleDownIcon = _interopRequireDefault(require("part:@sanity/base/angle-down-icon"));

var _react = _interopRequireDefault(require("react"));

var _Details = _interopRequireDefault(require("./Details.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Details extends _react.default.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "handleToggle", () => {
      this.setState(prevState => ({
        isOpen: !prevState.isOpen
      }));
    });

    this.state = {
      isOpen: props.isOpen
    };
  }

  render() {
    var _this$props = this.props,
        title = _this$props.title,
        children = _this$props.children;
    var isOpen = this.state.isOpen;
    return /*#__PURE__*/_react.default.createElement("button", {
      type: "button",
      className: _Details.default.root,
      "data-open": isOpen,
      onClick: this.handleToggle
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _Details.default.header,
      tabIndex: 0
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: _Details.default.iconContainer,
      "aria-hidden": "true",
      role: "img"
    }, /*#__PURE__*/_react.default.createElement(_angleDownIcon.default, null)), /*#__PURE__*/_react.default.createElement("span", {
      className: _Details.default.summary
    }, title)), /*#__PURE__*/_react.default.createElement("div", {
      className: _Details.default.content
    }, children));
  }

}

exports.default = Details;

_defineProperty(Details, "defaultProps", {
  title: 'Details',
  isOpen: false
});
