// @sanity/components/lib/layer/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _layer = require("part:@sanity/components/layer");

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function DefaultStory() {
  return /*#__PURE__*/_react.default.createElement(_layer.LayerProvider, null, /*#__PURE__*/_react.default.createElement(Example, null));
}

function Example() {
  var layer = (0, _layer.useLayer)();

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      open = _useState2[0],
      setOpen = _useState2[1];

  var handleOpen = (0, _react.useCallback)(() => setOpen(true), []);
  var handleClose = (0, _react.useCallback)(() => setOpen(false), []);
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("h1", null, "Layer example"), /*#__PURE__*/_react.default.createElement("pre", null, "depth=", layer.depth, ", size=", layer.size), /*#__PURE__*/_react.default.createElement("button", {
    onClick: handleOpen,
    type: "button"
  }, "Open"), open && /*#__PURE__*/_react.default.createElement(_layer.Layer, {
    style: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(Layer1, {
    onClose: handleClose
  })));
}

function Layer1(_ref) {
  var onClose = _ref.onClose;
  var layer = (0, _layer.useLayer)();

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      open = _useState4[0],
      setOpen = _useState4[1];

  var handleOpen = (0, _react.useCallback)(() => setOpen(true), []);
  var handleClose = (0, _react.useCallback)(() => setOpen(false), []);
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: '#fff',
      boxShadow: '0 5px 10px rgba(0, 0, 0, 0.25)',
      padding: 16
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      display: 'flex'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      flex: 1
    }
  }, /*#__PURE__*/_react.default.createElement("strong", null, "Layer")), /*#__PURE__*/_react.default.createElement("button", {
    onClick: onClose,
    type: "button"
  }, "\xD7")), /*#__PURE__*/_react.default.createElement("pre", null, "depth=", layer.depth, ", size=", layer.size), /*#__PURE__*/_react.default.createElement("button", {
    onClick: handleOpen,
    type: "button"
  }, "Open"), open && /*#__PURE__*/_react.default.createElement(_layer.Layer, {
    style: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(Layer2, {
    onClose: handleClose
  })));
}

function Layer2(_ref2) {
  var onClose = _ref2.onClose;
  var layer = (0, _layer.useLayer)();
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: '#fff',
      boxShadow: '0 5px 10px rgba(0, 0, 0, 0.25)',
      padding: 16
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      display: 'flex'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      flex: 1
    }
  }, /*#__PURE__*/_react.default.createElement("strong", null, "Layer")), /*#__PURE__*/_react.default.createElement("button", {
    onClick: onClose,
    type: "button"
  }, "\xD7")), /*#__PURE__*/_react.default.createElement("pre", null, "depth=", layer.depth, ", size=", layer.size));
}
