// @sanity/components/lib/snackbar/stories/stacked.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StackedStory = StackedStory;

var _react = _interopRequireDefault(require("react"));

var _item = _interopRequireDefault(require("part:@sanity/components/snackbar/item"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var globalDefaults = {
  offset: 10,
  isOpen: true,
  id: new Date().getTime() + Math.floor(Math.random()),
  setFocus: false,
  onClose: (0, _actions.action)('onClose()'),
  onDismiss: (0, _actions.action)('onDismiss()')
};

function StackedStory() {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_item.default, _extends({}, globalDefaults, {
    kind: "info",
    title: (0, _knobs.text)('title', 'This is a title placeholder', 'props'),
    isCloseable: (0, _knobs.boolean)('isCloseable', true, 'props'),
    icon: true,
    onSetHeight: (0, _actions.action)('onSetheight')
  })), /*#__PURE__*/_react.default.createElement(_item.default, _extends({}, globalDefaults, {
    offset: 75,
    kind: "warning",
    title: (0, _knobs.text)('title', 'This is a title placeholder', 'props'),
    isCloseable: (0, _knobs.boolean)('isCloseable', true, 'props'),
    icon: true,
    onSetHeight: (0, _actions.action)('onSetheight')
  })), /*#__PURE__*/_react.default.createElement(_item.default, _extends({}, globalDefaults, {
    offset: 140,
    kind: "success",
    title: (0, _knobs.text)('title', 'This is a title placeholder', 'props'),
    isCloseable: (0, _knobs.boolean)('onClose', true, 'props'),
    onSetHeight: (0, _actions.action)('onSetheight')
  })));
}
