// @sanity/components/lib/snackbar/stories/transitions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TransitionsStory = TransitionsStory;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var globalDefaults = {
  offset: 10,
  isOpen: true,
  id: new Date().getTime() + Math.floor(Math.random()),
  setFocus: false,
  onClose: (0, _actions.action)('onClose()'),
  onDismiss: (0, _actions.action)('onDismiss()')
}; // @todo: refactor to functional component

class SnackQueue extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "addToQueue", () => {
      var snack = this.props.snack;

      var newSnack = _objectSpread(_objectSpread({}, snack), {}, {
        id: new Date().getTime() + Math.floor(Math.random())
      });

      this.context.addToSnackQueue(newSnack);
    });
  }

  render() {
    (0, _knobs.button)('Add to queue', this.addToQueue, 'test');
    return /*#__PURE__*/_react.default.createElement("div", null);
  }

}

_defineProperty(SnackQueue, "contextTypes", {
  addToSnackQueue: _propTypes.default.func
});

function TransitionsStory() {
  var snack = _objectSpread(_objectSpread({}, globalDefaults), {}, {
    kind: (0, _knobs.select)('Kinds', ['info', 'success', 'warning', 'error'], 'info', 'props'),
    title: (0, _knobs.text)('Title', 'This is a title placeholder', 'props'),
    subtitle: (0, _knobs.text)('Subtitle', 'This is a subtitle placeholder', 'props'),
    setAutoFocus: (0, _knobs.boolean)('setAutoFocus', false, 'props'),
    isPersisted: (0, _knobs.boolean)('isPersisted', false, 'props'),
    autoDismissTimeout: (0, _knobs.number)('autoDismissTimeout (ms)', 4000, 'props'),
    icon: (0, _knobs.boolean)('Icon', false, 'props'),
    action: {
      title: (0, _knobs.text)('Action title', ''),
      callback: (0, _actions.action)('callback()')
    },
    isCloseable: (0, _knobs.boolean)('isCloseable', false, 'props'),
    onClose: (0, _actions.action)((0, _knobs.text)('onClose', 'Custom onClose', 'props')),
    onSetHeight: (0, _actions.action)((0, _knobs.text)('onSetHeight', 'onSetHeight', 'props')),
    children: (0, _knobs.text)('children', 'This is a children placeholder', 'props')
  });

  return /*#__PURE__*/_react.default.createElement(SnackQueue, {
    snack: snack
  });
}
