// @sanity/components/lib/snackbar/stories/withChildren.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WithChildrenStory = WithChildrenStory;

var _react = _interopRequireDefault(require("react"));

var _item = _interopRequireDefault(require("part:@sanity/components/snackbar/item"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var globalDefaults = {
  offset: 10,
  isOpen: true,
  id: new Date().getTime() + Math.floor(Math.random()),
  setFocus: false,
  onClose: (0, _actions.action)('onClose()'),
  onDismiss: (0, _actions.action)('onDismiss()')
};

function WithChildrenStory() {
  return /*#__PURE__*/_react.default.createElement(_item.default, _extends({}, globalDefaults, {
    kind: (0, _knobs.select)('Kind', ['success', 'error', 'warning', 'info'], 'info', 'props'),
    title: (0, _knobs.text)('Title', 'This is a title placeholder', 'props'),
    subtitle: (0, _knobs.text)('Subtitle', '', 'props'),
    icon: (0, _knobs.boolean)('Icon', false, 'props'),
    isCloseable: (0, _knobs.boolean)('isCloseable', true, 'props'),
    action: {
      title: (0, _knobs.text)('Action title', 'Action'),
      callback: (0, _actions.action)('callback()')
    },
    onClose: (0, _actions.action)((0, _knobs.text)('onClose', 'Custom onClose', 'props')),
    onSetHeight: (0, _actions.action)('onSetHeight')
  }), /*#__PURE__*/_react.default.createElement("div", null, (0, _knobs.text)('Children', 'This is the children placeholder', 'props')));
}
