// @sanity/components/lib/fileinput/stories/button.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonStory = ButtonStory;

var _react = _interopRequireDefault(require("react"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _button = _interopRequireDefault(require("part:@sanity/components/fileinput/button"));

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var centerStyle = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  height: '100%',
  width: '100%',
  position: 'absolute',
  top: 0,
  left: 0
};

function ButtonStory() {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/fileinput/button",
    propTables: [_button.default]
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    onSelect: (0, _actions.action)('onSelect')
  }, "Upload file\u2026")));
}
