// @sanity/components/lib/textfields/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _default = _interopRequireDefault(require("part:@sanity/components/textfields/default"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 640
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/textfields/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    label: (0, _knobs.text)('label', 'This is the label', 'props'),
    placeholder: (0, _knobs.text)('placeholder', 'This is the placeholder', 'props'),
    value: (0, _knobs.text)('value', undefined, 'props'),
    hasError: (0, _knobs.boolean)('hasError', false, 'props'),
    onChange: (0, _actions.action)('onChange'),
    onFocus: (0, _actions.action)('onFocus'),
    onClear: (0, _actions.action)('onClear'),
    isClearable: (0, _knobs.boolean)('isClearable', false, 'props') // hasFocus={boolean('hasFocus', false, 'props')}

  }))));
}
