// @sanity/field/lib/types/string/diff/StringFieldDiff.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StringFieldDiff = void 0;

var _react = _interopRequireDefault(require("react"));

var _diff = require("../../../diff");

var _StringPreview = require("../preview/StringPreview");

var _StringFieldDiff = _interopRequireDefault(require("./StringFieldDiff.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StringFieldDiff = (_ref) => {
  var diff = _ref.diff,
      schemaType = _ref.schemaType;
  var options = schemaType.options;

  if (options === null || options === void 0 ? void 0 : options.list) {
    // When the string is considered to be an "enum", don't show individual
    // string segment changes, rather treat is as a "from -> to" diff
    return /*#__PURE__*/_react.default.createElement(_diff.DiffFromTo, {
      cardClassName: _StringFieldDiff.default.card,
      diff: diff,
      previewComponent: _StringPreview.StringPreview,
      schemaType: schemaType
    });
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: _StringFieldDiff.default.root
  }, /*#__PURE__*/_react.default.createElement(_diff.DiffString, {
    diff: diff
  }));
};

exports.StringFieldDiff = StringFieldDiff;
