// @sanity/components/lib/previews/stories/card.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CardStory = CardStory;

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _linkIcon = _interopRequireDefault(require("part:@sanity/base/link-icon"));

var _card = _interopRequireDefault(require("part:@sanity/components/previews/card"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var renderMedia = () => {
  return /*#__PURE__*/_react.default.createElement("img", {
    src: "http://www.fillmurray.com/300/300",
    alt: "test"
  });
};

var renderStatus = () => {
  return /*#__PURE__*/_react.default.createElement("span", null, "Status ", /*#__PURE__*/_react.default.createElement(_linkIcon.default, null), " ", /*#__PURE__*/_react.default.createElement(_warningIcon.default, null));
};

var renderTitle = options => {
  return /*#__PURE__*/_react.default.createElement("span", null, "This ", /*#__PURE__*/_react.default.createElement("span", {
    style: {
      color: 'green'
    }
  }, "is"), " a ", /*#__PURE__*/_react.default.createElement("strong", null, "title"), "\xA0in the layout ", options.layout);
};

var renderSubtitle = () => {
  return /*#__PURE__*/_react.default.createElement("span", null, "This is a", ' ', /*#__PURE__*/_react.default.createElement("strong", {
    style: {
      color: 'red'
    }
  }, /*#__PURE__*/_react.default.createElement(_warningIcon.default, null), "subtitle"));
};

var renderDescription = () => {
  return /*#__PURE__*/_react.default.createElement("span", null, "This is the", ' ', /*#__PURE__*/_react.default.createElement("strong", {
    style: {
      color: 'red'
    }
  }, /*#__PURE__*/_react.default.createElement(_warningIcon.default, null), "description"));
};

var renderCustomChildren = () => {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'absolute',
      top: '0',
      left: '0',
      width: '100%',
      height: '100%',
      overflow: 'hidden',
      pointerEvents: 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'absolute',
      top: '0',
      right: '0'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'absolute',
      fontSize: '10px',
      textTransform: 'uppercase',
      top: '0',
      right: '0',
      fontWeight: 700,
      boxShadow: '0 0 5px rgba(0,0,0,0.2)',
      backgroundColor: 'yellow',
      padding: '0.2em 3em',
      transform: 'translate(28%, 43%) rotate(45deg)'
    }
  }, "New")));
};

var options = {
  functions: 'Functions',
  strings: 'Strings',
  elements: 'Element'
};

function CardStory() {
  var propType = (0, _knobs.select)('Type of props', options, 'strings');

  if (propType === 'functions') {
    return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
      style: {
        backgroundColor: 'none'
      }
    }, /*#__PURE__*/_react.default.createElement("div", {
      style: {
        width: '100%',
        maxWidth: 350
      }
    }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
      part: "part:@sanity/components/previews/card",
      propTables: [_card.default]
    }, /*#__PURE__*/_react.default.createElement(_card.default, {
      title: renderTitle,
      subtitle: renderSubtitle,
      description: renderDescription,
      status: renderStatus,
      media: renderMedia,
      isPlaceholder: (0, _knobs.boolean)('placeholder', true, 'props'),
      mediaDimensions: {
        width: 300,
        height: 225,
        aspect: (0, _knobs.number)('mediaDimensions.aspect', 4 / 3, 'props'),
        fit: 'crop'
      }
    }, (0, _knobs.boolean)('Custom children', false, 'test') && renderCustomChildren()))));
  }

  if (propType === 'elements') {
    return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
      style: {
        backgroundColor: 'none'
      }
    }, /*#__PURE__*/_react.default.createElement("div", {
      style: {
        width: '100%',
        maxWidth: 350
      }
    }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
      part: "part:@sanity/components/previews/card",
      propTables: [_card.default]
    }, /*#__PURE__*/_react.default.createElement(_card.default, {
      title: /*#__PURE__*/_react.default.createElement("span", null, "This ", /*#__PURE__*/_react.default.createElement("span", {
        style: {
          color: 'green'
        }
      }, "is"), " a ", /*#__PURE__*/_react.default.createElement("strong", null, "test")),
      subtitle: /*#__PURE__*/_react.default.createElement("span", null, "This is a ", /*#__PURE__*/_react.default.createElement("strong", {
        style: {
          color: 'red'
        }
      }, "subtitle")),
      description: /*#__PURE__*/_react.default.createElement("span", null, "This is the long the descriptions that should no be to long, beacuse we will cap it"),
      isPlaceholder: (0, _knobs.boolean)('placeholder', false, 'props'),
      media: (0, _knobs.boolean)('Show image', false) ? renderMedia : undefined,
      status: /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_linkIcon.default, null), /*#__PURE__*/_react.default.createElement(_warningIcon.default, null)),
      mediaDimensions: {
        width: 300,
        height: 225,
        aspect: (0, _knobs.number)('mediaDimensions.aspect', 4 / 3, 'props'),
        fit: 'crop'
      }
    }, (0, _knobs.boolean)('Custom children', false, 'test') && renderCustomChildren()))));
  }

  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
    style: {
      backgroundColor: 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 350
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/previews/card",
    propTables: [_card.default]
  }, /*#__PURE__*/_react.default.createElement(_card.default, {
    title: (0, _knobs.text)('title', 'Bill Murray', 'props'),
    subtitle: (0, _knobs.text)('subtitle', 'American actor', 'props'),
    description: (0, _knobs.text)('description', 'William James Murray is an American actor, comedian, and writer. He first gained exposure on Saturday Night Live, a series of performances that earned him his first Emmy Award, and later starred in.', 'props'),
    status: (0, _knobs.text)('status', '🔔 🐣 👻', 'props'),
    media: (0, _knobs.boolean)('Show image', true, 'test') ? renderMedia : undefined,
    isPlaceholder: (0, _knobs.boolean)('isPlaceholder', false, 'props'),
    mediaDimensions: {
      width: 300,
      height: 225,
      aspect: (0, _knobs.number)('mediaDimensions.aspect', 4 / 3, 'props'),
      fit: 'crop'
    }
  }, (0, _knobs.boolean)('Custom children', false) && renderCustomChildren()))));
}
