// @sanity/components/lib/previews/stories/createDocument.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateDocumentStory = CreateDocumentStory;

var _fileIcon = _interopRequireDefault(require("part:@sanity/base/file-icon"));

var _createDocument = _interopRequireDefault(require("part:@sanity/components/previews/create-document"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

var _components2 = require("./components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function CreateDocumentStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
    style: {
      backgroundColor: 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 350
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/previews/create-document",
    propTables: [_createDocument.default]
  }, /*#__PURE__*/_react.default.createElement(_components2.Stack, null, /*#__PURE__*/_react.default.createElement(_components2.PreviewCard, null, /*#__PURE__*/_react.default.createElement(_createDocument.default, {
    params: {
      intent: 'create',
      template: 'test'
    },
    title: (0, _knobs.text)('title', 'Movie', 'props'),
    subtitle: (0, _knobs.text)('subtitle', 'Sci-fi', 'props'),
    description: (0, _knobs.text)('description', 'Science fiction is a genre of speculative fiction that has been called the "literature of ideas". It typically deals with imaginative and futuristic concepts such as advanced science and technology, time travel, parallel universes, fictional worlds, space exploration, and extraterrestrial life.', 'props'),
    icon: _fileIcon.default
  }))))));
}
