// @sanity/components/lib/previews/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _warningIcon = _interopRequireDefault(require("part:@sanity/base/warning-icon"));

var _linkIcon = _interopRequireDefault(require("part:@sanity/base/link-icon"));

var _default = _interopRequireDefault(require("part:@sanity/components/previews/default"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

var _components2 = require("./components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var renderMedia = () => {
  return /*#__PURE__*/_react.default.createElement("img", {
    src: "http://www.fillmurray.com/300/300",
    alt: "test"
  });
};

var renderStatus = () => {
  return /*#__PURE__*/_react.default.createElement("span", null, "Status ", /*#__PURE__*/_react.default.createElement(_linkIcon.default, null), " ", /*#__PURE__*/_react.default.createElement(_warningIcon.default, null));
};

var renderTitle = options => {
  return /*#__PURE__*/_react.default.createElement("span", null, "This ", /*#__PURE__*/_react.default.createElement("span", {
    style: {
      color: 'green'
    }
  }, "is"), " a ", /*#__PURE__*/_react.default.createElement("strong", null, "title"), "\xA0in the layout ", options.layout);
};

var renderSubtitle = () => {
  return /*#__PURE__*/_react.default.createElement("span", null, "This is a", ' ', /*#__PURE__*/_react.default.createElement("strong", {
    style: {
      color: 'red'
    }
  }, /*#__PURE__*/_react.default.createElement(_warningIcon.default, null), "subtitle"));
};

var renderCustomChildren = () => {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'absolute',
      top: '0',
      left: '0',
      width: '100%',
      height: '100%',
      overflow: 'hidden',
      pointerEvents: 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'absolute',
      top: '0',
      right: '0'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'absolute',
      fontSize: '10px',
      textTransform: 'uppercase',
      top: '0',
      right: '0',
      fontWeight: 700,
      boxShadow: '0 0 5px rgba(0,0,0,0.2)',
      backgroundColor: 'yellow',
      padding: '0.2em 3em',
      transform: 'translate(28%, 43%) rotate(45deg)'
    }
  }, "New")));
};

var options = {
  functions: 'Functions',
  strings: 'Strings',
  elements: 'Element'
};

function DefaultStory() {
  var propType = (0, _knobs.select)('Type of props', options, 'strings');

  if (propType === 'functions') {
    return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
      style: {
        backgroundColor: 'none'
      }
    }, /*#__PURE__*/_react.default.createElement("div", {
      style: {
        width: '100%',
        maxWidth: 350
      }
    }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
      part: "part:@sanity/components/previews/default",
      propTables: [_default.default]
    }, /*#__PURE__*/_react.default.createElement(_components2.Stack, null, /*#__PURE__*/_react.default.createElement(_components2.PreviewCard, null, /*#__PURE__*/_react.default.createElement(_default.default, {
      title: renderTitle,
      subtitle: renderSubtitle,
      status: renderStatus,
      media: renderMedia,
      isPlaceholder: (0, _knobs.boolean)('placeholder', false, 'props'),
      progress: (0, _knobs.number)('progress', undefined, {
        range: true,
        min: 0,
        max: 100,
        step: 1
      }, 'props')
    }, (0, _knobs.boolean)('Custom children', false) && renderCustomChildren()))))));
  }

  if (propType === 'elements') {
    return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
      style: {
        backgroundColor: 'none'
      }
    }, /*#__PURE__*/_react.default.createElement("div", {
      style: {
        width: '100%',
        maxWidth: 350
      }
    }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
      part: "part:@sanity/components/previews/default",
      propTables: [_default.default]
    }, /*#__PURE__*/_react.default.createElement(_components2.Stack, null, /*#__PURE__*/_react.default.createElement(_components2.PreviewCard, null, /*#__PURE__*/_react.default.createElement(_default.default, {
      title: /*#__PURE__*/_react.default.createElement("span", null, "This ", /*#__PURE__*/_react.default.createElement("span", {
        style: {
          color: 'green'
        }
      }, "is"), " a ", /*#__PURE__*/_react.default.createElement("strong", null, "test")),
      subtitle: /*#__PURE__*/_react.default.createElement("span", null, "This is a ", /*#__PURE__*/_react.default.createElement("strong", {
        style: {
          color: 'red'
        }
      }, "subtitle")),
      isPlaceholder: (0, _knobs.boolean)('placeholder', false, 'props'),
      media: (0, _knobs.boolean)('Show image', true, 'test') ? renderMedia : undefined,
      status: /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_linkIcon.default, null), /*#__PURE__*/_react.default.createElement(_warningIcon.default, null)),
      progress: (0, _knobs.number)('progress', undefined, {
        range: true,
        min: 0,
        max: 100,
        step: 1
      }, 'props')
    }, (0, _knobs.boolean)('Custom children', false, 'test') && renderCustomChildren()))))));
  }

  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
    style: {
      backgroundColor: 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 350
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/previews/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement(_components2.Stack, null, /*#__PURE__*/_react.default.createElement(_components2.PreviewCard, null, /*#__PURE__*/_react.default.createElement(_default.default, {
    title: (0, _knobs.text)('title', 'This is the title an it is very long, so long that it should be ellipsed', 'props'),
    subtitle: (0, _knobs.text)('subtitle', "This is the title an it is very long, so long that it should be ellipsed.\n             This is the title an it is very long, so long that it should be ellipsed", 'props'),
    status: (0, _knobs.text)('status', '🔔 🐣 👻', 'props'),
    media: (0, _knobs.boolean)('Show image', true, 'props') ? renderMedia : undefined,
    isPlaceholder: (0, _knobs.boolean)('placeholder', false, 'props'),
    progress: (0, _knobs.number)('progress', undefined, {
      range: true,
      min: 0,
      max: 100,
      step: 1
    }, 'props')
  }, (0, _knobs.boolean)('Custom children', false, 'test') && renderCustomChildren()))))));
}
