// @sanity/components/lib/previews/stories/inline.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InlineStory = InlineStory;

var _inline = _interopRequireDefault(require("part:@sanity/components/previews/inline"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var renderMedia = () => {
  return /*#__PURE__*/_react.default.createElement("img", {
    src: "http://www.fillmurray.com/300/300",
    alt: "test"
  });
};

var renderTitle = options => {
  return /*#__PURE__*/_react.default.createElement("span", null, "This ", /*#__PURE__*/_react.default.createElement("span", {
    style: {
      color: 'green'
    }
  }, "is"), " a ", /*#__PURE__*/_react.default.createElement("strong", null, "title"), "\xA0in the layout ", options.layout);
};

var options = {
  functions: 'Functions',
  strings: 'Strings',
  elements: 'Element'
};

function InlineStory() {
  var propType = (0, _knobs.select)('Type of props', options, 'strings');

  if (propType === 'functions') {
    return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
      part: "part:@sanity/components/previews/inline",
      propTables: [_inline.default]
    }, /*#__PURE__*/_react.default.createElement("p", null, "This is a text, and suddenly a inline preview appearst before", /*#__PURE__*/_react.default.createElement(_inline.default, {
      title: renderTitle,
      media: renderMedia
    }, (0, _knobs.boolean)('Custom children', false, 'test') && /*#__PURE__*/_react.default.createElement("span", null, "This is custom children")), "this word.")));
  }

  if (propType === 'elements') {
    return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
      part: "part:@sanity/components/previews/inline",
      propTables: [_inline.default]
    }, /*#__PURE__*/_react.default.createElement("p", null, "This is a text, and suddenly a inline preview appearst before", /*#__PURE__*/_react.default.createElement(_inline.default, {
      title: /*#__PURE__*/_react.default.createElement("span", null, "title"),
      media: renderMedia
    }, (0, _knobs.boolean)('Custom children', false, 'test') && /*#__PURE__*/_react.default.createElement("span", null, "This is custom children")), "this word.")));
  }

  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/previews/inline",
    propTables: [_inline.default]
  }, /*#__PURE__*/_react.default.createElement("p", null, "This is a text, and suddenly an inline preview appears before", ' ', /*#__PURE__*/_react.default.createElement(_inline.default, {
    title: (0, _knobs.text)('title', 'This is the title', 'props'),
    media: renderMedia
  }, (0, _knobs.boolean)('Custom children', false, 'test') && /*#__PURE__*/_react.default.createElement("span", null, "This is custom children")), ' ', "this word.")));
}
