// @sanity/components/lib/previews/stories/media.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MediaStory = MediaStory;

var _media = _interopRequireDefault(require("part:@sanity/components/previews/media"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

var _components2 = require("./components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var renderMedia = () => {
  return /*#__PURE__*/_react.default.createElement("img", {
    src: "http://www.fillmurray.com/300/300",
    alt: "test"
  });
};

var renderCustomChildren = () => {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'absolute',
      top: '0',
      left: '0',
      width: '100%',
      height: '100%',
      overflow: 'hidden',
      pointerEvents: 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'absolute',
      top: '0',
      right: '0'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'absolute',
      fontSize: '10px',
      textTransform: 'uppercase',
      top: '0',
      right: '0',
      fontWeight: 700,
      boxShadow: '0 0 5px rgba(0,0,0,0.2)',
      backgroundColor: 'yellow',
      padding: '0.2em 3em',
      transform: 'translate(28%, 43%) rotate(45deg)'
    }
  }, "New")));
};

function MediaStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
    style: {
      backgroundColor: 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      maxWidth: 350
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/previews/media",
    propTables: [_media.default]
  }, /*#__PURE__*/_react.default.createElement(_components2.Stack, null, /*#__PURE__*/_react.default.createElement(_components2.PreviewCard, null, /*#__PURE__*/_react.default.createElement(_media.default, {
    title: (0, _knobs.text)('title (prop)', 'This is the title', 'props'),
    subtitle: (0, _knobs.text)('subtitle (prop)', 'This is the subtitle', 'props'),
    description: (0, _knobs.text)('description', 'This is the long the descriptions that should no be to long, beacuse we will cap it', 'props'),
    media: renderMedia,
    isPlaceholder: (0, _knobs.boolean)('isplaceholder', false, 'props')
  }, (0, _knobs.boolean)('Custom children', false, 'test') && renderCustomChildren()))))));
}
