// @sanity/components/lib/autocomplete/stories/defaultWithSuggestions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultWithSuggestionsStory = DefaultWithSuggestionsStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _chance = _interopRequireDefault(require("chance"));

var _default = _interopRequireDefault(require("part:@sanity/components/autocomplete/default"));

var _storybook = require("part:@sanity/storybook");

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

function DefaultWithSuggestionsStory() {
  var suggestions = (0, _range2.default)(10).map((item, i) => {
    return {
      id: "".concat(i),
      title: chance.name()
    };
  });
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement("form", {
    style: {
      width: '100%',
      maxWidth: '30em'
    }
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    value: "One",
    suggestions: suggestions // isOpen
    ,
    label: "Autocomplete component",
    onChange: (0, _storybook.action)('onChange')
  })));
}
