// @sanity/components/lib/validation/stories/status.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StatusStory = StatusStory;

var _range2 = _interopRequireDefault(require("lodash/range"));

var _react = _interopRequireDefault(require("react"));

var _chance = _interopRequireDefault(require("chance"));

var _status = _interopRequireDefault(require("part:@sanity/components/validation/status"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _components = require("part:@sanity/storybook/components");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

var mockMarkers = function mockMarkers() {
  var length = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 5;
  return (0, _range2.default)(length).map((_, i) => ({
    type: 'validation',
    level: 'error',
    path: ["test_".concat(i)],
    item: {
      message: chance.sentence(),
      paths: [],

      cloneWithMessage() {
        throw new Error('nope');
      }

    }
  }));
};

function StatusStory() {
  var markers = mockMarkers();
  var hideTooltip = (0, _knobs.boolean)('Hide tooltip', false, 'Props');
  var showSummary = (0, _knobs.boolean)('Show summary', false, 'Props');
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/validation/status",
    propTables: [_status.default]
  }, /*#__PURE__*/_react.default.createElement(_status.default, {
    hideTooltip: hideTooltip,
    markers: markers,
    showSummary: showSummary
  })));
}
