// @sanity/desk-tool/lib/components/paneItem/PaneItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PaneItem;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _schema = _interopRequireDefault(require("part:@sanity/base/schema"));

var _preview = require("part:@sanity/base/preview");

var _folderIcon = _interopRequireDefault(require("part:@sanity/base/folder-icon"));

var _fileIcon = _interopRequireDefault(require("part:@sanity/base/file-icon"));

var _DocumentPaneItemPreview = _interopRequireDefault(require("../../components/DocumentPaneItemPreview"));

var _getIconWithFallback = _interopRequireDefault(require("../../utils/getIconWithFallback"));

var _MissingSchemaType = _interopRequireDefault(require("../../components/MissingSchemaType"));

var _PaneItemWrapper = _interopRequireDefault(require("./PaneItemWrapper"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function PaneItem(props) {
  var id = props.id,
      isSelected = props.isSelected,
      schemaType = props.schemaType,
      layout = props.layout,
      icon = props.icon,
      value = props.value;
  var useGrid = layout === 'card' || layout === 'media';

  var hasSchemaType = schemaType && schemaType.name && _schema.default.get(schemaType.name);

  var preview;

  if (value && value._id) {
    preview = hasSchemaType ? /*#__PURE__*/_react.default.createElement(_DocumentPaneItemPreview.default, {
      icon: (0, _getIconWithFallback.default)(icon, schemaType, _fileIcon.default),
      layout: layout,
      schemaType: schemaType,
      value: value
    }) : /*#__PURE__*/_react.default.createElement(_MissingSchemaType.default, {
      value: value
    });
  } else {
    preview = /*#__PURE__*/_react.default.createElement(_preview.SanityDefaultPreview, {
      icon: (0, _getIconWithFallback.default)(icon, schemaType, _folderIcon.default),
      layout: layout,
      value: value
    });
  }

  return /*#__PURE__*/_react.default.createElement(_PaneItemWrapper.default, {
    id: id,
    isSelected: isSelected,
    layout: layout,
    useGrid: useGrid
  }, preview);
}

PaneItem.propTypes = {
  id: _propTypes.default.string.isRequired,
  layout: _propTypes.default.string,
  isSelected: _propTypes.default.bool,
  icon: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.func]),
  value: _propTypes.default.shape({
    _id: _propTypes.default.string,
    _type: _propTypes.default.string,
    title: _propTypes.default.string,
    subtitle: _propTypes.default.string,
    media: _propTypes.default.oneOfType([_propTypes.default.node, _propTypes.default.func])
  }),
  schemaType: _propTypes.default.shape({
    name: _propTypes.default.string,
    icon: _propTypes.default.func
  })
};
PaneItem.defaultProps = {
  layout: 'default',
  icon: undefined,
  value: null,
  isSelected: false,
  schemaType: null
};
