// @sanity/components/lib/popover/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _popover = require("part:@sanity/components/popover");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  var children = (0, _knobs.text)('Children', 'Content', 'Props');
  var open = (0, _knobs.boolean)('Open', true, 'Props');
  var placement = (0, _knobs.select)('Placement', {
    'top-start': 'Top start',
    top: 'Top',
    'top-end': 'Top end',
    'right-start': 'Right start',
    right: 'Right',
    'right-end': 'Right end',
    'bottom-start': 'Bottom start',
    bottom: 'Bottom',
    'bottom-end': 'Bottom end',
    'left-start': 'Left start',
    left: 'Left',
    'left-end': 'Left end'
  }, 'bottom', 'Props');
  var popoverContent = children ? /*#__PURE__*/_react.default.createElement("div", {
    style: {
      padding: '1em'
    }
  }, children) : null;
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/dialogs/default",
    propTables: [_popover.Popover]
  }, /*#__PURE__*/_react.default.createElement(_popover.Popover, {
    content: popoverContent,
    open: open,
    placement: placement
  }, /*#__PURE__*/_react.default.createElement("span", {
    style: {
      display: 'inline-block',
      padding: '0.25em',
      background: 'rgba(127, 127, 127, 0.5)'
    }
  }, "Reference"))));
}
