// @sanity/components/lib/tabs/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _tab = _interopRequireDefault(require("part:@sanity/components/tabs/tab"));

var _tabPanel = _interopRequireDefault(require("part:@sanity/components/tabs/tab-panel"));

var _tabList = _interopRequireDefault(require("part:@sanity/components/tabs/tab-list"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  var currentTabId = (0, _knobs.select)('Current tab', ['tab-edit', 'tab-seo', 'tab-preview'], 'tab-edit', 'props');
  var tabs = [{
    id: 'tab-edit',
    label: 'Edit'
  }, {
    id: 'tab-seo',
    label: 'SEO'
  }, {
    id: 'tab-preview',
    label: 'Preview'
  }];
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: 'white'
    }
  }, /*#__PURE__*/_react.default.createElement(_tabList.default, null, tabs.map(tab => /*#__PURE__*/_react.default.createElement(_tab.default, {
    "aria-controls": "tab-panel",
    id: tab.id,
    isActive: tab.id === currentTabId,
    key: tab.id,
    label: tab.label,
    onClick: (0, _actions.action)('onClick')
  }))), /*#__PURE__*/_react.default.createElement(_tabPanel.default, {
    "aria-labelledby": currentTabId,
    id: "tab-panel"
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      borderTop: '1px solid #ddd',
      padding: '1em'
    }
  }, "Content: ", currentTabId)));
}
