// @sanity/components/lib/tabs/stories/many.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ManyStory = ManyStory;

var _tab = _interopRequireDefault(require("part:@sanity/components/tabs/tab"));

var _tabPanel = _interopRequireDefault(require("part:@sanity/components/tabs/tab-panel"));

var _tabList = _interopRequireDefault(require("part:@sanity/components/tabs/tab-list"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ManyStory() {
  var currentTabId = (0, _knobs.select)('Current tab', ['tab-edit', 'tab-seo', 'tab-preview'], 'tab-edit', 'props');
  var tabs = [{
    id: 'tab-0',
    label: 'Foo'
  }, {
    id: 'tab-1',
    label: 'Foo'
  }, {
    id: 'tab-2',
    label: 'Foo'
  }, {
    id: 'tab-3',
    label: 'Foo'
  }, {
    id: 'tab-4',
    label: 'Foo'
  }, {
    id: 'tab-5',
    label: 'Foo'
  }, {
    id: 'tab-6',
    label: 'Foo'
  }, {
    id: 'tab-7',
    label: 'Foo'
  }, {
    id: 'tab-8',
    label: 'Foo'
  }, {
    id: 'tab-9',
    label: 'Foo'
  }, {
    id: 'tab-10',
    label: 'Foo'
  }, {
    id: 'tab-11',
    label: 'Foo'
  }, {
    id: 'tab-12',
    label: 'Foo'
  }, {
    id: 'tab-13',
    label: 'Foo'
  }, {
    id: 'tab-14',
    label: 'Foo'
  }, {
    id: 'tab-15',
    label: 'Foo'
  }, {
    id: 'tab-16',
    label: 'Foo'
  }, {
    id: 'tab-17',
    label: 'Foo'
  }, {
    id: 'tab-18',
    label: 'Foo'
  }, {
    id: 'tab-19',
    label: 'Foo'
  }, {
    id: 'tab-20',
    label: 'Foo'
  }];
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: 'white'
    }
  }, /*#__PURE__*/_react.default.createElement(_tabList.default, null, tabs.map(tab => /*#__PURE__*/_react.default.createElement(_tab.default, {
    "aria-controls": "tab-panel",
    id: tab.id,
    isActive: tab.id === currentTabId,
    key: tab.id,
    label: tab.label,
    onClick: (0, _actions.action)('onClick')
  }))), /*#__PURE__*/_react.default.createElement(_tabPanel.default, {
    "aria-labelledby": currentTabId,
    id: "tab-panel"
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      borderTop: '1px solid #eee',
      padding: '1em'
    }
  }, "Content: ", currentTabId)));
}
