// @sanity/components/lib/loading/stories/spinner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpinnerStory = SpinnerStory;

var _react = _interopRequireDefault(require("react"));

var _sanityLogoIcon = _interopRequireDefault(require("part:@sanity/base/sanity-logo-icon"));

var _spinner = _interopRequireDefault(require("part:@sanity/components/loading/spinner"));

var _spinnerStyle = _interopRequireDefault(require("part:@sanity/components/loading/spinner-style"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function renderCustomChildren() {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _spinnerStyle.default.message
  }, /*#__PURE__*/_react.default.createElement(_sanityLogoIcon.default, null), " :wave:");
}

function SpinnerStory() {
  return /*#__PURE__*/_react.default.createElement(_components.CenteredContainer, {
    style: {
      background: 'none'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '100%',
      minWidth: 300,
      maxWidth: 500,
      background: '#fff',
      lineHeight: 0
    }
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/loading/spinner",
    propTables: [_spinner.default]
  }, /*#__PURE__*/_react.default.createElement(_spinner.default, {
    inline: (0, _knobs.boolean)('inline', false, 'props'),
    message: (0, _knobs.text)('Message', 'This is the message', 'props'),
    fullscreen: (0, _knobs.boolean)('fullscreen', false, 'props'),
    center: (0, _knobs.boolean)('center', false, 'props')
  }, (0, _knobs.boolean)('Custom children', false) && renderCustomChildren()))));
}
