// @sanity/form-builder/lib/inputs/FileInput/FileInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _partition4 = _interopRequireDefault(require("lodash/partition"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _ChangeIndicator = require("@sanity/base/lib/change-indicators/ChangeIndicator");

var _changeIndicators = require("@sanity/base/lib/change-indicators");

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _button = _interopRequireDefault(require("part:@sanity/components/fileinput/button"));

var _circle = _interopRequireDefault(require("part:@sanity/components/progress/circle"));

var _editIcon = _interopRequireDefault(require("part:@sanity/base/edit-icon"));

var _visibilityIcon = _interopRequireDefault(require("part:@sanity/base/visibility-icon"));

var _fileIcon = _interopRequireDefault(require("part:@sanity/base/file-icon"));

var _uploadIcon = _interopRequireDefault(require("part:@sanity/base/upload-icon"));

var _fullscreen = _interopRequireDefault(require("part:@sanity/components/dialogs/fullscreen"));

var _buttonGrid = _interopRequireDefault(require("part:@sanity/components/buttons/button-grid"));

var _anchor = _interopRequireDefault(require("part:@sanity/components/buttons/anchor"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/snackbar/default"));

var _UploadTargetFieldset = _interopRequireDefault(require("../../utils/UploadTargetFieldset"));

var _WithMaterializedReference = _interopRequireDefault(require("../../utils/WithMaterializedReference"));

var _PatchEvent = _interopRequireWildcard(require("../../PatchEvent"));

var _FormBuilderInput = require("../../FormBuilderInput");

var _UploadPlaceholder = _interopRequireDefault(require("../common/UploadPlaceholder"));

var _FileInput = _interopRequireDefault(require("./FileInput.css"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var HIDDEN_FIELDS = ['asset', 'hotspot', 'crop'];

class FileInput extends _react.default.PureComponent {
  constructor() {
    super(...arguments);

    _defineProperty(this, "_focusArea", void 0);

    _defineProperty(this, "uploadSubscription", void 0);

    _defineProperty(this, "state", {
      isUploading: false,
      isAdvancedEditOpen: false,
      uploadError: null,
      hasFocus: false
    });

    _defineProperty(this, "handleRemoveButtonClick", () => {
      var getValuePath = this.context.getValuePath;
      var value = this.props.value;
      var parentPathSegment = getValuePath().slice(-1)[0]; // String path segment mean an object path, while a number or a
      // keyed segment means we're a direct child of an array

      var isArrayElement = typeof parentPathSegment !== 'string'; // When removing the file, _type and _key are "meta"-properties and
      // are not significant unless other properties are present. Thus, we
      // want to remove the entire "container" object if these are the only
      // properties present, BUT only if we're not an array element, as
      // removing the array element will close the selection dialog. Instead,
      // when closing the dialog, the array logic will check for an "empty"
      // value and remove it for us

      var allKeys = Object.keys(value);
      var remainingKeys = allKeys.filter(key => !['_type', '_key', '_upload', 'asset'].includes(key));
      var isEmpty = remainingKeys.length === 0;
      var removeKeys = ['asset'].concat(allKeys.filter(key => ['_upload'].includes(key))).map(key => (0, _PatchEvent.unset)([key]));
      this.props.onChange(_PatchEvent.default.from(isEmpty && !isArrayElement ? (0, _PatchEvent.unset)() : removeKeys));
    });

    _defineProperty(this, "handleCancelUpload", () => {
      this.cancelUpload();
    });

    _defineProperty(this, "handleSelectFile", files => {
      this.uploadFirstAccepted(files);
    });

    _defineProperty(this, "renderMaterializedAsset", assetDocument => {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: _FileInput.default.previewAsset
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: _FileInput.default.fileIcon
      }, /*#__PURE__*/_react.default.createElement(_fileIcon.default, null)), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
        className: _FileInput.default.fileLabel
      }, assetDocument.originalFilename, " "), /*#__PURE__*/_react.default.createElement(_anchor.default, {
        href: "".concat(assetDocument.url, "?dl"),
        inverted: true,
        size: "small"
      }, "Download")));
    });

    _defineProperty(this, "handleFieldChange", (event, field) => {
      var _this$props = this.props,
          onChange = _this$props.onChange,
          type = _this$props.type;
      onChange(event.prefixAll(field.name).prepend((0, _PatchEvent.setIfMissing)({
        _type: type.name
      })));
    });

    _defineProperty(this, "handleStartAdvancedEdit", () => {
      this.setState({
        isAdvancedEditOpen: true
      });
    });

    _defineProperty(this, "handleStopAdvancedEdit", () => {
      this.setState({
        isAdvancedEditOpen: false
      });
    });

    _defineProperty(this, "handleClearUploadError", () => {
      this.setState({
        uploadError: null
      });
    });

    _defineProperty(this, "handleFocus", event => {
      this.setState({
        hasFocus: true
      });
      this.props.onFocus(event);
    });

    _defineProperty(this, "handleBlur", () => {
      this.setState({
        hasFocus: false
      });
      this.props.onBlur();
    });

    _defineProperty(this, "setFocusArea", el => {
      this._focusArea = el;
    });

    _defineProperty(this, "getUploadOptions", file => {
      var _this$props2 = this.props,
          type = _this$props2.type,
          resolveUploader = _this$props2.resolveUploader;
      var uploader = resolveUploader && resolveUploader(type, file);
      return uploader ? [{
        type: type,
        uploader
      }] : [];
    });

    _defineProperty(this, "handleUpload", (_ref) => {
      var file = _ref.file,
          uploader = _ref.uploader;
      this.uploadWith(uploader, file);
    });
  }

  clearUploadStatus() {
    this.props.onChange(_PatchEvent.default.from([(0, _PatchEvent.unset)(['_upload'])])); // todo: this is kind of hackish
  }

  cancelUpload() {
    if (this.uploadSubscription) {
      this.uploadSubscription.unsubscribe();
      this.clearUploadStatus();
    }
  }

  uploadFirstAccepted(fileList) {
    var _this$props3 = this.props,
        resolveUploader = _this$props3.resolveUploader,
        type = _this$props3.type;
    var match;
    Array.from(fileList).some(file => {
      var uploader = resolveUploader(type, file);

      if (uploader) {
        match = {
          file,
          uploader
        };
        return true;
      }

      return false;
    });

    if (match) {
      this.uploadWith(match.uploader, match.file);
    }
  }

  uploadWith(uploader, file) {
    var _this$props4 = this.props,
        type = _this$props4.type,
        onChange = _this$props4.onChange;
    var options = {
      metadata: (0, _get2.default)(type, 'options.metadata'),
      storeOriginalFilename: (0, _get2.default)(type, 'options.storeOriginalFilename')
    };
    this.cancelUpload();
    this.setState({
      isUploading: true
    });
    onChange(_PatchEvent.default.from([(0, _PatchEvent.setIfMissing)({
      _type: type.name
    })]));
    this.uploadSubscription = uploader.upload(file, type, options).subscribe({
      next: uploadEvent => {
        if (uploadEvent.patches) {
          onChange(_PatchEvent.default.from(uploadEvent.patches));
        }
      },
      error: err => {
        this.setState({
          uploadError: err
        });
        this.clearUploadStatus();
      },
      complete: () => {
        onChange(_PatchEvent.default.from([(0, _PatchEvent.unset)(['hotspot']), (0, _PatchEvent.unset)(['crop'])]));
        this.setState({
          isUploading: false
        });
      }
    });
  }

  renderUploadState(uploadState) {
    var isUploading = this.state.isUploading;
    var isComplete = uploadState.progress === 100;
    var filename = (0, _get2.default)(uploadState, 'file.name');
    return /*#__PURE__*/_react.default.createElement("div", {
      className: _FileInput.default.uploadState
    }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_circle.default, {
      percent: status === 'complete' ? 100 : uploadState.progress,
      text: isComplete ? 'Complete' : "Uploading".concat(filename ? " \"".concat(filename, "\"") : '...'),
      completed: isComplete,
      showPercent: true,
      animation: true
    })), /*#__PURE__*/_react.default.createElement("div", {
      className: _FileInput.default.cancelButton
    }, isUploading && /*#__PURE__*/_react.default.createElement(_default.default, {
      kind: "simple",
      color: "danger",
      onClick: this.handleCancelUpload
    }, "Cancel"))));
  }

  renderAdvancedEdit(fields) {
    return /*#__PURE__*/_react.default.createElement(_fullscreen.default, {
      title: "Edit details",
      onClose: this.handleStopAdvancedEdit,
      isOpen: true
    }, /*#__PURE__*/_react.default.createElement("div", null, this.renderFields(fields)));
  } // eslint-disable-next-line class-methods-use-this


  renderSelectFileButton() {
    // Single asset source (just a normal button)
    // @todo add select handling here
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      inverted: true
    }, "Select");
  }

  renderFields(fields) {
    return fields.map(field => this.renderField(field));
  }

  renderField(field) {
    var _this$props5 = this.props,
        value = _this$props5.value,
        level = _this$props5.level,
        focusPath = _this$props5.focusPath,
        onFocus = _this$props5.onFocus,
        readOnly = _this$props5.readOnly,
        onBlur = _this$props5.onBlur,
        presence = _this$props5.presence;
    var fieldValue = value && value[field.name];
    return /*#__PURE__*/_react.default.createElement(_FormBuilderInput.FormBuilderInput, {
      key: field.name,
      value: fieldValue,
      type: field.type,
      onChange: ev => this.handleFieldChange(ev, field),
      path: [field.name],
      onFocus: onFocus,
      onBlur: onBlur,
      readOnly: readOnly || field.type.readOnly,
      focusPath: focusPath,
      level: level,
      presence: presence
    });
  }

  renderAsset() {
    var _this$props6 = this.props,
        value = _this$props6.value,
        materialize = _this$props6.materialize,
        readOnly = _this$props6.readOnly;

    if (value && value.asset) {
      return /*#__PURE__*/_react.default.createElement(_WithMaterializedReference.default, {
        reference: value.asset,
        materialize: materialize
      }, this.renderMaterializedAsset);
    }

    return readOnly ? /*#__PURE__*/_react.default.createElement("span", null, "Field is read only") : /*#__PURE__*/_react.default.createElement(_UploadPlaceholder.default, {
      hasFocus: this.state.hasFocus
    });
  }

  focus() {
    if (this._focusArea) {
      this._focusArea.focus();
    }
  }

  render() {
    var _value$asset, _compareValue$asset;

    var _this$props7 = this.props,
        type = _this$props7.type,
        value = _this$props7.value,
        compareValue = _this$props7.compareValue,
        level = _this$props7.level,
        markers = _this$props7.markers,
        readOnly = _this$props7.readOnly,
        presence = _this$props7.presence;
    var _this$state = this.state,
        isAdvancedEditOpen = _this$state.isAdvancedEditOpen,
        uploadError = _this$state.uploadError,
        hasFocus = _this$state.hasFocus;

    var _partition2 = (0, _partition4.default)(type.fields.filter(field => !HIDDEN_FIELDS.includes(field.name)), 'type.options.isHighlighted'),
        _partition3 = _slicedToArray(_partition2, 2),
        highlightedFields = _partition3[0],
        otherFields = _partition3[1];

    var accept = (0, _get2.default)(type, 'options.accept', '');
    var hasAsset = value && value.asset;
    var isInside = presence.map(item => {
      var otherFieldsPath = otherFields.map(field => field.name);
      return item.path.some(path => otherFieldsPath.includes(path)) ? item.identity : null;
    }).filter(String);
    return /*#__PURE__*/_react.default.createElement(_UploadTargetFieldset.default, {
      markers: markers,
      legend: type.title,
      description: type.description,
      level: level,
      onFocus: this.handleFocus,
      onBlur: this.handleBlur,
      onUpload: this.handleUpload,
      getUploadOptions: this.getUploadOptions,
      ref: this.setFocusArea,
      presence: presence.filter(item => item.path[0] === '$' || isInside.includes(item.identity)),
      changeIndicator: false
    }, uploadError && /*#__PURE__*/_react.default.createElement(_default2.default, {
      kind: "error",
      isPersisted: true,
      actionTitle: "OK",
      onAction: this.handleClearUploadError,
      title: "Upload error",
      subtitle: /*#__PURE__*/_react.default.createElement("div", null, "We're really sorry, but the upload could not be completed.")
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)(_FileInput.default.root, readOnly && _FileInput.default.readOnly, hasFocus && _FileInput.default.focused)
    }, /*#__PURE__*/_react.default.createElement(_ChangeIndicator.ChangeIndicatorCompareValueProvider, {
      value: value === null || value === void 0 ? void 0 : (_value$asset = value.asset) === null || _value$asset === void 0 ? void 0 : _value$asset._ref,
      compareValue: compareValue === null || compareValue === void 0 ? void 0 : (_compareValue$asset = compareValue.asset) === null || _compareValue$asset === void 0 ? void 0 : _compareValue$asset._ref
    }, /*#__PURE__*/_react.default.createElement(_changeIndicators.ChangeIndicator, null, /*#__PURE__*/_react.default.createElement("div", {
      className: _FileInput.default.content
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _FileInput.default.assetWrapper
    }, value && value._upload && /*#__PURE__*/_react.default.createElement("div", {
      className: _FileInput.default.uploadState
    }, this.renderUploadState(value._upload)), this.renderAsset())))), /*#__PURE__*/_react.default.createElement("div", {
      className: _FileInput.default.functions
    }, /*#__PURE__*/_react.default.createElement(_buttonGrid.default, null, !readOnly && /*#__PURE__*/_react.default.createElement(_button.default, {
      inverted: true,
      icon: _uploadIcon.default,
      onSelect: this.handleSelectFile,
      accept: accept
    }, "Upload"), value && otherFields.length > 0 && /*#__PURE__*/_react.default.createElement(_default.default, {
      icon: readOnly ? _visibilityIcon.default : _editIcon.default,
      inverted: true,
      title: readOnly ? 'View details' : 'Edit details',
      onClick: this.handleStartAdvancedEdit
    }, readOnly ? 'View details' : 'Edit'), !readOnly && hasAsset && /*#__PURE__*/_react.default.createElement(_default.default, {
      color: "danger",
      inverted: true,
      onClick: this.handleRemoveButtonClick
    }, "Remove"), isAdvancedEditOpen && this.renderAdvancedEdit(otherFields)))), highlightedFields.length > 0 && this.renderFields(highlightedFields));
  }

}

exports.default = FileInput;

_defineProperty(FileInput, "contextTypes", {
  getValuePath: _propTypes.default.func
});
