// @sanity/base/lib/hooks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "useDocumentPresence", {
  enumerable: true,
  get: function get() {
    return _hooks.useDocumentPresence;
  }
});
Object.defineProperty(exports, "useGlobalPresence", {
  enumerable: true,
  get: function get() {
    return _hooks.useGlobalPresence;
  }
});
Object.defineProperty(exports, "useUser", {
  enumerable: true,
  get: function get() {
    return _hooks2.useUser;
  }
});
Object.defineProperty(exports, "useUserColor", {
  enumerable: true,
  get: function get() {
    return _hooks3.useUserColor;
  }
});
Object.defineProperty(exports, "useTimeAgo", {
  enumerable: true,
  get: function get() {
    return _useTimeAgo.useTimeAgo;
  }
});
Object.defineProperty(exports, "useDocumentValues", {
  enumerable: true,
  get: function get() {
    return _useDocumentValues.useDocumentValues;
  }
});

var _hooks = require("./datastores/presence/hooks");

var _hooks2 = require("./datastores/user/hooks");

var _hooks3 = require("./user-color/hooks");

var _useTimeAgo = require("./time/useTimeAgo");

var _useDocumentValues = require("./datastores/document/useDocumentValues");
