// @sanity/components/lib/panes/stories/pane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PaneStory = PaneStory;

var _composeIcon = _interopRequireDefault(require("part:@sanity/base/compose-icon"));

var _trashIcon = _interopRequireDefault(require("part:@sanity/base/trash-icon"));

var _default = _interopRequireDefault(require("part:@sanity/components/buttons/default"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/panes/default"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

var _pane = _interopRequireDefault(require("./pane.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var menuItems = [{
  action: 'first',
  title: 'First item',
  key: '1'
}, {
  action: 'second',
  title: 'Second item',
  key: '2'
}, {
  action: 'third',
  title: 'Third item',
  key: '3',
  params: {
    some: 'param'
  }
}, {
  action: 'fourth',
  title: 'Extra item',
  key: '4',
  group: 'extra'
}, {
  title: 'Add',
  key: '5',
  group: 'actions',
  icon: _composeIcon.default,
  showAsAction: true,
  action: (0, _actions.action)('Add')
}];

var renderActions = () => {
  return [/*#__PURE__*/_react.default.createElement(_default.default, {
    color: "primary",
    icon: _composeIcon.default,
    key: "add",
    kind: "simple",
    padding: "small",
    title: "Add"
  }), /*#__PURE__*/_react.default.createElement(_default.default, {
    color: "danger",
    icon: _trashIcon.default,
    key: "delete",
    kind: "simple",
    padding: "small",
    title: "Delete"
  })];
};

function PaneStory() {
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/panes/default",
    propTables: [_default2.default]
  }, /*#__PURE__*/_react.default.createElement(_components.DebugRouterProvider, null, /*#__PURE__*/_react.default.createElement(_default2.default, {
    index: 0,
    title: (0, _knobs.text)('title', 'This is the default pane title', 'props'),
    isCollapsed: (0, _knobs.boolean)('isCollapsed', false, 'props'),
    onExpand: (0, _actions.action)('onExpand'),
    onCollapse: (0, _actions.action)('onCollapse'),
    onAction: (0, _actions.action)('onAction'),
    menuItems: menuItems,
    renderActions: renderActions
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _pane.default.root
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _pane.default.text
  }, "Contents")))));
}
