// @sanity/components/lib/portal/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _portal = require("part:@sanity/components/portal");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/dialogs/default",
    propTables: [_portal.Portal]
  }, /*#__PURE__*/_react.default.createElement(PortalExample, null));
}

function PortalExample() {
  var rootRef = (0, _react.useRef)(null);
  var portalRef = (0, _react.useRef)(document.createElement('div'));
  (0, _react.useEffect)(() => {
    if (rootRef.current) {
      rootRef.current.appendChild(portalRef.current);
    }

    return () => {
      if (rootRef.current) {
        rootRef.current.removeChild(portalRef.current);
      }
    };
  }, []);
  return /*#__PURE__*/_react.default.createElement(_portal.PortalProvider, {
    element: portalRef.current
  }, /*#__PURE__*/_react.default.createElement("div", {
    ref: rootRef
  }, "Portal demo:"), /*#__PURE__*/_react.default.createElement(_portal.Portal, null, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: 'rgba(255, 0, 0, 0.1)',
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      background: '#fff',
      padding: 20
    }
  }, "Hello from portal"))));
}
