// @sanity/desk-tool/lib/panes/documentPane/documentOperationResults.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentOperationResults = void 0;

var _reactHooks = require("@sanity/react-hooks");

var _default = _interopRequireDefault(require("part:@sanity/components/snackbar/default"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getOpErrorTitle(op) {
  if (op === 'delete') {
    return "An error occurred while attempting to delete this document.\n      This usually means that there are other documents that refers to it.";
  }

  if (op === 'unpublish') {
    return "An error occurred while attempting to unpublish this document.\n      This usually means that there are other documents that refers to it.";
  }

  return "An error occurred during ".concat(op);
}

function getOpSuccessTitle(op) {
  if (op === 'publish') {
    return "This document is now published.";
  }

  if (op === 'unpublish') {
    return "This document is now unpublished. A draft has been created from the latest published version.";
  }

  if (op === 'discardChanges') {
    return "All changes since last publish has now been discarded. The discarded draft can still be recovered from history";
  }

  if (op === 'delete') {
    return "This document is now deleted. It can still be recovered from history and if you continue editing it will be recreated.";
  }

  return "Successfully performed ".concat(op, " on document");
}

var IGNORE_OPS = ['patch', 'commit'];

var DocumentOperationResults = /*#__PURE__*/_react.default.memo(props => {
  var event = (0, _reactHooks.useDocumentOperationEvent)(props.id, props.type);

  if (!event) {
    return null;
  }

  if (event && event.type === 'error') {
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      kind: "error",
      key: Math.random(),
      title: getOpErrorTitle(event.op),
      subtitle: /*#__PURE__*/_react.default.createElement("details", null, /*#__PURE__*/_react.default.createElement("summary", null, "Details"), event.error.message)
    });
  }

  if (event && event.type === 'success' && !IGNORE_OPS.includes(event.op)) {
    return /*#__PURE__*/_react.default.createElement(_default.default, {
      key: Math.random(),
      kind: "success",
      title: getOpSuccessTitle(event.op)
    });
  }

  return null;
});

exports.DocumentOperationResults = DocumentOperationResults;
DocumentOperationResults.displayName = 'DocumentOperationResults';
