// @sanity/components/lib/fieldsets/stories/collapsible.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CollapsibleStory = CollapsibleStory;

var _chance = _interopRequireDefault(require("chance"));

var _default = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/textfields/default"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

function CollapsibleStory() {
  var level = (0, _knobs.number)('start level', 1);
  var isCollapsible = (0, _knobs.boolean)('isCollapsible', true);
  return /*#__PURE__*/_react.default.createElement(_components.Container, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/fieldsets/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      margin: '0 auto',
      maxWidth: 640
    }
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    isCollapsible: isCollapsible,
    legend: "Noooo, pick me, pick me!",
    description: chance.sentence(),
    level: level
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 1
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 1,
    description: "Test this one"
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 1
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 1
  })), /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: "Fieldsets can be collapsed",
    description: chance.sentence(),
    level: level
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    isCollapsible: isCollapsible,
    legend: "Open me",
    description: chance.sentence(),
    level: level + 1
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2,
    description: "Test this one"
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2
  })), /*#__PURE__*/_react.default.createElement(_default.default, {
    isCollapsible: isCollapsible,
    legend: "No, open me!",
    description: chance.sentence(),
    level: level + 1
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2,
    description: "Test this one"
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2
  })), /*#__PURE__*/_react.default.createElement(_default.default, {
    isCollapsible: isCollapsible,
    legend: "Noooo, pick me, pick me!",
    description: chance.sentence(),
    level: level + 1
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2,
    description: "Test this one"
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2
  }), /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Content",
    level: level + 2
  }))))));
}
