// @sanity/components/lib/fieldsets/stories/columns.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColumnsStory = ColumnsStory;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/formfields/default"));

var _default3 = _interopRequireDefault(require("part:@sanity/components/textinputs/default"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _chance = _interopRequireDefault(require("chance"));

var _components = require("part:@sanity/storybook/components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

function DummyField(_ref) {
  var level = _ref.level;
  return /*#__PURE__*/_react.default.createElement(_default2.default, {
    label: "Label",
    description: "Description",
    level: level,
    inline: false
  }, /*#__PURE__*/_react.default.createElement(_default3.default, {
    value: ""
  }));
}

function ColumnsStory() {
  var columns = (0, _knobs.number)('columns', 3);
  var level = (0, _knobs.number)('start level', 1);
  return /*#__PURE__*/_react.default.createElement(_components.Container, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/fieldsets/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      margin: '0 auto',
      maxWidth: 640
    }
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: "".concat(columns, "columns"),
    columns: columns,
    description: chance.paragraph(),
    level: level
  }, /*#__PURE__*/_react.default.createElement(DummyField, {
    level: level + 1
  }), /*#__PURE__*/_react.default.createElement(DummyField, {
    level: level + 1
  }), /*#__PURE__*/_react.default.createElement(DummyField, {
    level: level + 1
  }), /*#__PURE__*/_react.default.createElement(DummyField, {
    level: level + 1
  }), /*#__PURE__*/_react.default.createElement(DummyField, {
    level: level + 1
  }), /*#__PURE__*/_react.default.createElement(DummyField, {
    level: level + 1
  })))));
}
