// @sanity/components/lib/fieldsets/stories/deeplyNested.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeeplyNestedStory = DeeplyNestedStory;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _chance = _interopRequireDefault(require("chance"));

var _components = require("part:@sanity/storybook/components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

function DeeplyNestedStory() {
  return /*#__PURE__*/_react.default.createElement(_components.Container, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/fieldsets/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      margin: '0 auto',
      maxWidth: 640
    }
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: "Dude, I heard you like fieldsets\u2026",
    description: chance.paragraph(),
    level: 1,
    tabIndex: 0
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: "So I put a fieldset in a fieldset\u2026",
    description: chance.paragraph(),
    level: 2,
    tabIndex: 0
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: "In a fieldset\u2026",
    description: chance.paragraph(),
    level: 3,
    tabIndex: 0
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: "In a fieldset!",
    description: chance.paragraph(),
    level: 4,
    tabIndex: 0
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: "In a fieldset!",
    description: chance.paragraph(),
    level: 5,
    tabIndex: 0
  }, "Content"))))))));
}
