// @sanity/components/lib/fieldsets/stories/default.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultStory = DefaultStory;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _default2 = _interopRequireDefault(require("part:@sanity/components/previews/default"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _components = require("part:@sanity/storybook/components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function DefaultStory() {
  return /*#__PURE__*/_react.default.createElement(_components.Container, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/fieldsets/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      margin: '0 auto',
      maxWidth: 640
    }
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: (0, _knobs.text)('legend', 'This is the legend', 'props'),
    description: (0, _knobs.text)('description', 'This is the description', 'props'),
    level: (0, _knobs.number)('level', 1, 'props'),
    transparent: (0, _knobs.boolean)('transparent', false, 'props'),
    tabIndex: (0, _knobs.boolean)('tabIndex', false, 'test') ? 0 : -1,
    markers: (0, _knobs.boolean)('show markers', false, 'test') ? [] : undefined
  }, /*#__PURE__*/_react.default.createElement(_default2.default, {
    media: /*#__PURE__*/_react.default.createElement("img", {
      src: "http://www.fillmurray.com/300/300",
      alt: "test"
    }),
    title: "Long title alksdmaldsm aslkdma ldkamsldk amsldkasm dlkmasdlkams ldkamdlk amsdlaskmdlaks dmalskd maslkdm aslkdmaslk dmalskdm aslkdml"
  })))));
}
