// @sanity/components/lib/fieldsets/stories/nested.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NestedStory = NestedStory;

var _react = _interopRequireDefault(require("react"));

var _default = _interopRequireDefault(require("part:@sanity/components/fieldsets/default"));

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _chance = _interopRequireDefault(require("chance"));

var _components = require("part:@sanity/storybook/components");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chance = new _chance.default();

function NestedStory() {
  return /*#__PURE__*/_react.default.createElement(_components.Container, null, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/fieldsets/default",
    propTables: [_default.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      margin: '0 auto',
      maxWidth: 640
    }
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: "This is the legend",
    description: "Fieldsets supports beeing inside itselfs, and get new styling!",
    level: 1
  }, /*#__PURE__*/_react.default.createElement(_default.default, {
    legend: "This is the legend in a nested fieldset",
    description: chance.paragraph(),
    level: 2
  }, "A nested fieldset")))));
}
