// @sanity/components/lib/selects/stories/styleSelect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyleSelectStory = StyleSelectStory;

var _react = _interopRequireDefault(require("react"));

var _storybook = require("part:@sanity/storybook");

var _style = _interopRequireDefault(require("part:@sanity/components/selects/style"));

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var styleItems = [{
  title: 'Paragraph',
  key: 'style-paragraph'
}, {
  title: 'Heading 1',
  key: 'style-heading1'
}, {
  title: 'Heading 2',
  key: 'style-heading2'
}, {
  title: 'Heading 3',
  key: 'style-heading3'
}, {
  title: 'Heading 4',
  key: 'style-heading4'
}, {
  title: 'Heading 5',
  key: 'style-heading5'
}];
var centerStyle = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  height: '100%',
  width: '100%',
  position: 'absolute',
  top: 0,
  left: 0
};

var renderStyleItem = function renderStyleItem(item) {
  switch (item.key) {
    case 'style-paragraph':
      return /*#__PURE__*/_react.default.createElement("div", {
        style: {
          fontSize: '1em',
          fontWeight: 'normal'
        }
      }, item.title);

    case 'style-heading1':
      return /*#__PURE__*/_react.default.createElement("div", {
        style: {
          fontSize: '2em',
          fontWeight: 'bold'
        }
      }, item.title);

    case 'style-heading2':
      return /*#__PURE__*/_react.default.createElement("div", {
        style: {
          fontSize: '1.5em',
          fontWeight: 'bold'
        }
      }, item.title);

    case 'style-heading3':
      return /*#__PURE__*/_react.default.createElement("div", {
        style: {
          fontSize: '1.2em',
          fontWeight: 'bold'
        }
      }, item.title);

    default:
      return /*#__PURE__*/_react.default.createElement("div", null, "Style: ", item.title);
  }
};

function StyleSelectStory() {
  return /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/selects/style",
    propTables: [_style.default]
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_style.default, {
    placeholder: (0, _knobs.text)('placeholder', 'This is the placeholder', 'props'),
    transparent: (0, _knobs.boolean)('transparent', false, 'props'),
    onChange: (0, _storybook.action)('onChange'),
    onOpen: (0, _storybook.action)('onOpen'),
    padding: (0, _knobs.select)('Padding', ['large', 'default', 'small', 'none'], 'default', 'props'),
    renderItem: renderStyleItem,
    items: styleItems
  })));
}
