// @sanity/form-builder/lib/FormBuilderErrorBoundary.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Details = _interopRequireDefault(require("./inputs/common/Details"));

var _FormBuilderErrorBoundary = _interopRequireDefault(require("./styles/FormBuilderErrorBoundary.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

// eslint-disable-next-line react/require-optimization
class ErrorBoundary extends _react.default.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "ref", /*#__PURE__*/_react.default.createRef());

    this.state = {
      error: null
    };
  }

  static getDerivedStateFromError(error) {
    return {
      error
    };
  } // eslint-disable-next-line class-methods-use-this


  componentDidCatch(error) {
    // eslint-disable-next-line no-console
    console.error('Input component crashed:', error);
  }

  render() {
    var error = this.state.error;

    if (!error) {
      return this.props.children;
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: _FormBuilderErrorBoundary.default.root,
      tabIndex: 0,
      ref: this.ref
    }, /*#__PURE__*/_react.default.createElement("h3", null, "Rats! The input component crashed."), /*#__PURE__*/_react.default.createElement(_Details.default, null, /*#__PURE__*/_react.default.createElement("pre", null, error.message), /*#__PURE__*/_react.default.createElement("em", null, "(check developer tools console for details)")));
  }

}

var _default = ErrorBoundary;
exports.default = _default;
