// @sanity/server/lib/configs/webpack.config.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _webpack = _interopRequireDefault(require("webpack"));

var _applyStaticLoaderFix = _interopRequireDefault(require("../util/applyStaticLoaderFix"));

var _webpack2 = _interopRequireDefault(require("./webpack.config"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = config => {
  var baseConfig = (0, _webpack2.default)(config);
  return Object.assign({}, baseConfig, {
    devtool: 'cheap-module-source-map',
    output: Object.assign({
      pathinfo: true
    }, baseConfig.output),
    entry: Object.assign({}, baseConfig.entry, {
      vendor: [require.resolve('eventsource-polyfill'), require.resolve('../browser/hot-client')].concat(baseConfig.entry.vendor)
    }),
    resolve: {
      alias: Object.assign({}, baseConfig.resolve.alias, {
        'react-dom': require.resolve('@hot-loader/react-dom'),
        'webpack-hot-middleware/client': require.resolve('../browser/hot-client')
      }),
      extensions: baseConfig.resolve.extensions
    },
    plugins: (baseConfig.plugins || []).concat([new _webpack.default.HotModuleReplacementPlugin(), new _webpack.default.NoEmitOnErrorsPlugin()]),
    module: Object.assign({}, baseConfig.module, {
      rules: (0, _applyStaticLoaderFix.default)(baseConfig, config)
    })
  });
};

exports.default = _default;
