// @sanity/components/lib/tags/stories/tags.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TagsStory = TagsStory;

var _textfield = _interopRequireDefault(require("part:@sanity/components/tags/textfield"));

var _actions = require("part:@sanity/storybook/addons/actions");

var _knobs = require("part:@sanity/storybook/addons/knobs");

var _sanity = _interopRequireDefault(require("part:@sanity/storybook/addons/sanity"));

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var centerStyle = {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  height: '100%',
  width: '100%',
  position: 'absolute',
  top: 0,
  left: 0
};

function TagsStory() {
  var tags = ['Test', 'Sanity'];
  return /*#__PURE__*/_react.default.createElement("div", {
    style: centerStyle
  }, /*#__PURE__*/_react.default.createElement(_sanity.default, {
    part: "part:@sanity/components/tags/textfield",
    propTables: [_textfield.default]
  }, /*#__PURE__*/_react.default.createElement(_textfield.default, {
    label: (0, _knobs.text)('label', 'Tags', 'props'),
    readOnly: (0, _knobs.boolean)('readOnly', false, 'props'),
    placeholder: (0, _knobs.text)('placeholder', 'This is the placeholder', 'props'),
    value: (0, _knobs.array)('value', tags, 'props'),
    onChange: (0, _actions.action)('onChange')
  })));
}
