// @sanity/base/lib/change-indicators/helpers/getOffsetsTo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _scrollUtils = require("./scrollUtils");

var getOffsetsTo = (source, target) => {
  var el = source;
  var bounds = {
    top: 0,
    left: 0,
    height: Number.MAX_SAFE_INTEGER,
    width: Number.MAX_SAFE_INTEGER
  };
  var top = 0;
  var left = 0;
  var foundScrollContainer = false;

  while (el && el !== target) {
    if (foundScrollContainer) {
      bounds.top += el.offsetTop;
      bounds.left += el.offsetLeft;
    }

    if ((0, _scrollUtils.hasOverflowScroll)(el)) {
      bounds.top = el.offsetTop;
      bounds.height = el.offsetHeight;
      bounds.left = el.offsetLeft;
      bounds.width = el.offsetWidth;
      foundScrollContainer = true;
    }

    top += el.offsetTop - el.scrollTop;
    left += el.offsetLeft - el.scrollLeft;
    el = el.offsetParent;
  }

  return {
    rect: {
      top,
      left,
      height: source.offsetHeight,
      width: source.offsetWidth
    },
    bounds: bounds
  };
};

var time = (label, fn) => {
  return function () {
    // eslint-disable-next-line no-console
    console.time(label);
    var ret = fn(...arguments); // eslint-disable-next-line no-console

    console.timeEnd(label);
    return ret;
  };
};

var _default = getOffsetsTo;
exports.default = _default;
